{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aaaf3cc6-37e8-43e4-a57e-d9f3d868ca68",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "from sqlite3 import Error\n",
    "from sqlite3 import IntegrityError\n",
    "from ipyfilechooser import FileChooser\n",
    "import pandas as pd\n",
    "import io\n",
    "import requests\n",
    "import numpy\n",
    "from scipy import stats\n",
    "import math\n",
    "from scipy.stats import ttest_ind\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d047872c-f454-4b21-ae5f-fbe76ded5a3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4b10652523cb49ccb6e11197e07bcfbd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FileChooser(path='/Users/Joseph/Desktop', filename='', title='', show_hidden=False, select_desc='Select', chan…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "starting_directory = '/Users/Joseph/Desktop'\n",
    "baseline_chooser = FileChooser(starting_directory)\n",
    "display(baseline_chooser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0cf95420-e48d-4d6d-964f-dbbc41cc0e08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "423eb65c4a794e8fbb58b2bd60b2662b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FileChooser(path='/Users/Joseph/Desktop', filename='', title='', show_hidden=False, select_desc='Select', chan…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tap_chooser=FileChooser(starting_directory)\n",
    "display(tap_chooser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ad41b49-c54a-41bc-b2a4-12865e70cb27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/Joseph/Desktop/MWT_Datasets\n"
     ]
    }
   ],
   "source": [
    "Screen='Glia_Genes_Screen'\n",
    "folder_path=baseline_chooser.selected_path\n",
    "print(folder_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8bdae795-4e5b-4371-acd2-40d77be2796a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Time   n  Number  Instantaneous Speed  Interval Speed   Bias   Width  \\\n",
      "0  490.037  52      41               0.0319          0.0285  0.075  0.0906   \n",
      "1  490.077  52      41               0.0361          0.0343  0.075  0.0922   \n",
      "2  490.115  52      41               0.0323          0.0333  0.075  0.0935   \n",
      "3  490.154  52      41               0.0335          0.0290  0.075  0.0915   \n",
      "4  490.197  52      41               0.0337          0.0317  0.050  0.0909   \n",
      "\n",
      "   Length      Area  Angular Speed  Aspect Ratio  Kink  Curve    Crab  \\\n",
      "0  0.8885  0.100122            1.9         0.262  40.5   26.4  0.0065   \n",
      "1  0.8920  0.101651            2.1         0.264  40.1   26.4  0.0052   \n",
      "2  0.8905  0.102931            2.2         0.262  39.9   26.2  0.0040   \n",
      "3  0.8868  0.100922            2.5         0.259  40.4   26.3  0.0049   \n",
      "4  0.8880  0.100193            2.4         0.254  41.1   26.7  0.0044   \n",
      "\n",
      "  Plate_id      Date             Screen dataset Gene Allele  \n",
      "0    B1210  20220919  Glia_Genes_Screen      N2   N2     N2  \n",
      "1    B1210  20220919  Glia_Genes_Screen      N2   N2     N2  \n",
      "2    B1210  20220919  Glia_Genes_Screen      N2   N2     N2  \n",
      "3    B1210  20220919  Glia_Genes_Screen      N2   N2     N2  \n",
      "4    B1210  20220919  Glia_Genes_Screen      N2   N2     N2  \n"
     ]
    }
   ],
   "source": [
    "baseline_output = pd.read_csv(baseline_chooser.selected, index_col=0).drop(columns=['index'])\n",
    "print(baseline_output.head())\n",
    "# X=baseline_output.drop(columns=['Time','n','Number','Tap','Puff','x','y','plate','dataset','Gene','Allele'])\n",
    "# print(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a3581dfb-50a2-41f2-87d2-5ca53439c229",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      time  dura   dist      prob     speed  plate      Date Plate_id  \\\n",
      "0  599.993  2.67  0.529  0.857143  0.198127      1  20220919    B1210   \n",
      "1  609.924  2.05  0.479  0.852941  0.233659      1  20220919    B1210   \n",
      "2  619.982  1.86  0.386  0.488372  0.207527      1  20220919    B1210   \n",
      "3  629.946  1.87  0.371  0.500000  0.198396      1  20220919    B1210   \n",
      "4  639.927  1.94  0.384  0.382979  0.197938      1  20220919    B1210   \n",
      "\n",
      "              Screen  taps dataset Gene Allele  \n",
      "0  Glia_Genes_Screen     1      N2   N2     N2  \n",
      "1  Glia_Genes_Screen     2      N2   N2     N2  \n",
      "2  Glia_Genes_Screen     3      N2   N2     N2  \n",
      "3  Glia_Genes_Screen     4      N2   N2     N2  \n",
      "4  Glia_Genes_Screen     5      N2   N2     N2  \n"
     ]
    }
   ],
   "source": [
    "tap_output = pd.read_csv(tap_chooser.selected, index_col=0)\n",
    "print(tap_output.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5684c2d7-7f9e-4a9e-921e-3532535b2c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test=tap_output.dropna()\n",
    "# print(test[test.dataset=='mod-5_n822'].taps.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a9cec40a-82bb-4783-a0a9-fa6d9fcfa6d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        time  dura   dist      prob     speed  plate      Date Plate_id  \\\n",
      "0    599.993  2.67  0.529  0.857143  0.198127      1  20220919    B1210   \n",
      "1    599.980  2.32  0.525  0.944444  0.226293      2  20220811    C0811   \n",
      "2    599.986  2.82  0.686  0.857143  0.243262      3  20220811    A0811   \n",
      "3    599.991  2.73  0.698  0.958333  0.255678      4  20220811    C0811   \n",
      "4    599.979  2.99  0.798  0.875000  0.266890      5  20220811    A0811   \n",
      "..       ...   ...    ...       ...       ...    ...       ...      ...   \n",
      "559  599.976  2.69  0.761  0.818182  0.282900      5  20221105    A1105   \n",
      "560  599.991  2.63  0.730  0.916667  0.277567      6  20221105    B1105   \n",
      "561  599.986  2.88  0.772  0.833333  0.268056      7  20221105    C1105   \n",
      "562  599.961  2.97  0.701  0.842105  0.236027      8  20221105    C1105   \n",
      "563  599.979  2.92  0.802  0.875000  0.274658      9  20221105    B1105   \n",
      "\n",
      "                Screen  taps      dataset   Gene Allele  \n",
      "0    Glia_Genes_Screen     1           N2     N2     N2  \n",
      "1    Glia_Genes_Screen     1           N2     N2     N2  \n",
      "2    Glia_Genes_Screen     1           N2     N2     N2  \n",
      "3    Glia_Genes_Screen     1           N2     N2     N2  \n",
      "4    Glia_Genes_Screen     1           N2     N2     N2  \n",
      "..                 ...   ...          ...    ...    ...  \n",
      "559  Glia_Genes_Screen     1  vab-3_ok452  vab-3  ok452  \n",
      "560  Glia_Genes_Screen     1  vab-3_ok452  vab-3  ok452  \n",
      "561  Glia_Genes_Screen     1  vab-3_ok452  vab-3  ok452  \n",
      "562  Glia_Genes_Screen     1  vab-3_ok452  vab-3  ok452  \n",
      "563  Glia_Genes_Screen     1  vab-3_ok452  vab-3  ok452  \n",
      "\n",
      "[564 rows x 13 columns]\n",
      "-------------------------------------------------------\n",
      "         time  dura   dist      prob     speed  plate      Date Plate_id  \\\n",
      "0     869.997  1.06  0.135  0.027027  0.127358      1  20220919    B1210   \n",
      "1     879.949  1.32  0.278  0.051282  0.210606      1  20220919    B1210   \n",
      "2     889.945  1.22  0.220  0.063830  0.180328      1  20220919    B1210   \n",
      "3     869.976  1.31  0.335  0.210526  0.255725      2  20220811    C0811   \n",
      "4     879.974  0.55  0.142  0.294118  0.258182      2  20220811    C0811   \n",
      "...       ...   ...    ...       ...       ...    ...       ...      ...   \n",
      "1684  879.996  0.75  0.189  0.130435  0.252000      8  20221105    C1105   \n",
      "1685  889.986  0.22  0.038  0.136364  0.172727      8  20221105    C1105   \n",
      "1686  869.954  0.47  0.083  0.225806  0.176596      9  20221105    B1105   \n",
      "1687  879.957  0.56  0.097  0.230769  0.173214      9  20221105    B1105   \n",
      "1688  889.920  1.10  0.268  0.162791  0.243636      9  20221105    B1105   \n",
      "\n",
      "                 Screen  taps      dataset   Gene Allele  \n",
      "0     Glia_Genes_Screen    28           N2     N2     N2  \n",
      "1     Glia_Genes_Screen    29           N2     N2     N2  \n",
      "2     Glia_Genes_Screen    30           N2     N2     N2  \n",
      "3     Glia_Genes_Screen    28           N2     N2     N2  \n",
      "4     Glia_Genes_Screen    29           N2     N2     N2  \n",
      "...                 ...   ...          ...    ...    ...  \n",
      "1684  Glia_Genes_Screen    29  vab-3_ok452  vab-3  ok452  \n",
      "1685  Glia_Genes_Screen    30  vab-3_ok452  vab-3  ok452  \n",
      "1686  Glia_Genes_Screen    28  vab-3_ok452  vab-3  ok452  \n",
      "1687  Glia_Genes_Screen    29  vab-3_ok452  vab-3  ok452  \n",
      "1688  Glia_Genes_Screen    30  vab-3_ok452  vab-3  ok452  \n",
      "\n",
      "[1689 rows x 13 columns]\n",
      "-------------------------------------------------------\n",
      "-------------------------------------------------------\n",
      "PD_final_taps\n",
      "         dataset      Date Plate_id             Screen   Gene Allele  plate  \\\n",
      "0             N2  20220419    A0419  Glia_Genes_Screen     N2     N2     38   \n",
      "1             N2  20220419    A0419  Glia_Genes_Screen     N2     N2     40   \n",
      "2             N2  20220419    B0419  Glia_Genes_Screen     N2     N2     41   \n",
      "3             N2  20220419    B0419  Glia_Genes_Screen     N2     N2     42   \n",
      "4             N2  20220419    C0419  Glia_Genes_Screen     N2     N2     39   \n",
      "..           ...       ...      ...                ...    ...    ...    ...   \n",
      "559  vab-3_ok452  20221105    C1105  Glia_Genes_Screen  vab-3  ok452      7   \n",
      "560  vab-3_ok452  20221105    C1105  Glia_Genes_Screen  vab-3  ok452      8   \n",
      "561  vab-3_ok452  20221205    B1205  Glia_Genes_Screen  vab-3  ok452      1   \n",
      "562  vab-3_ok452  20221205    C1205  Glia_Genes_Screen  vab-3  ok452      2   \n",
      "563  vab-3_ok452  20221205    C1205  Glia_Genes_Screen  vab-3  ok452      3   \n",
      "\n",
      "           time  final_dura      dist  final_prob  final_speed  taps  \n",
      "0    879.986333    1.423333  0.384000    0.148742     0.271807  29.0  \n",
      "1    879.986333    1.090000  0.248667    0.237841     0.222914  29.0  \n",
      "2    879.989667    1.636667  0.282000    0.235411     0.173263  29.0  \n",
      "3    879.964000    1.260000  0.237000    0.186952     0.188915  29.0  \n",
      "4    879.949333    1.803333  0.377000    0.239351     0.207652  29.0  \n",
      "..          ...         ...       ...         ...          ...   ...  \n",
      "559  879.981333    0.566667  0.122333    0.300000     0.213481  29.0  \n",
      "560  879.992000    0.560000  0.129333    0.149539     0.217163  29.0  \n",
      "561  879.946333    1.056667  0.215000    0.269312     0.203101  29.0  \n",
      "562  879.972667    0.890000  0.215667    0.311877     0.241052  29.0  \n",
      "563  879.994667    0.910000  0.189667    0.263399     0.203795  29.0  \n",
      "\n",
      "[564 rows x 13 columns]\n",
      "-------------------------------------------------------\n",
      "PD_habit_levels\n",
      "      time_x  init_dura  dist_x  init_prob  init_speed  plate      Date  \\\n",
      "0    599.993       2.67   0.529   0.857143    0.198127      1  20220919   \n",
      "1    599.980       2.32   0.525   0.944444    0.226293      2  20220811   \n",
      "2    599.986       2.82   0.686   0.857143    0.243262      3  20220811   \n",
      "3    599.991       2.73   0.698   0.958333    0.255678      4  20220811   \n",
      "4    599.979       2.99   0.798   0.875000    0.266890      5  20220811   \n",
      "..       ...        ...     ...        ...         ...    ...       ...   \n",
      "559  599.976       2.69   0.761   0.818182    0.282900      5  20221105   \n",
      "560  599.991       2.63   0.730   0.916667    0.277567      6  20221105   \n",
      "561  599.986       2.88   0.772   0.833333    0.268056      7  20221105   \n",
      "562  599.961       2.97   0.701   0.842105    0.236027      8  20221105   \n",
      "563  599.979       2.92   0.802   0.875000    0.274658      9  20221105   \n",
      "\n",
      "    Plate_id             Screen  taps_x      dataset   Gene Allele  \\\n",
      "0      B1210  Glia_Genes_Screen       1           N2     N2     N2   \n",
      "1      C0811  Glia_Genes_Screen       1           N2     N2     N2   \n",
      "2      A0811  Glia_Genes_Screen       1           N2     N2     N2   \n",
      "3      C0811  Glia_Genes_Screen       1           N2     N2     N2   \n",
      "4      A0811  Glia_Genes_Screen       1           N2     N2     N2   \n",
      "..       ...                ...     ...          ...    ...    ...   \n",
      "559    A1105  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452   \n",
      "560    B1105  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452   \n",
      "561    C1105  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452   \n",
      "562    C1105  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452   \n",
      "563    B1105  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452   \n",
      "\n",
      "         time_y  final_dura    dist_y  final_prob  final_speed  taps_y  \n",
      "0    879.963667    1.200000  0.211000    0.047380     0.172764    29.0  \n",
      "1    879.975000    0.833333  0.210667    0.316363     0.252032    29.0  \n",
      "2    879.976667    1.356667  0.323333    0.271435     0.240509    29.0  \n",
      "3    879.982000    0.783333  0.191667    0.291783     0.235854    29.0  \n",
      "4    879.981333    0.876667  0.209000    0.192593     0.244481    29.0  \n",
      "..          ...         ...       ...         ...          ...     ...  \n",
      "559  879.985333    0.330000  0.064667    0.304584     0.236088    29.0  \n",
      "560  879.980667    0.853333  0.229333    0.218081     0.239567    29.0  \n",
      "561  879.981333    0.566667  0.122333    0.300000     0.213481    29.0  \n",
      "562  879.992000    0.560000  0.129333    0.149539     0.217163    29.0  \n",
      "563  879.943667    0.710000  0.149333    0.206455     0.197815    29.0  \n",
      "\n",
      "[551 rows x 19 columns]\n",
      "-------------------------------------------------------\n",
      "final version PD_habit_levels\n",
      "     init_dura  init_prob  init_speed  plate      Date Plate_id  \\\n",
      "0         2.67   0.857143    0.198127      1  20220919    B1210   \n",
      "1         2.32   0.944444    0.226293      2  20220811    C0811   \n",
      "2         2.82   0.857143    0.243262      3  20220811    A0811   \n",
      "3         2.73   0.958333    0.255678      4  20220811    C0811   \n",
      "4         2.99   0.875000    0.266890      5  20220811    A0811   \n",
      "..         ...        ...         ...    ...       ...      ...   \n",
      "559       2.69   0.818182    0.282900      5  20221105    A1105   \n",
      "560       2.63   0.916667    0.277567      6  20221105    B1105   \n",
      "561       2.88   0.833333    0.268056      7  20221105    C1105   \n",
      "562       2.97   0.842105    0.236027      8  20221105    C1105   \n",
      "563       2.92   0.875000    0.274658      9  20221105    B1105   \n",
      "\n",
      "                Screen  taps_x      dataset   Gene Allele  final_dura  \\\n",
      "0    Glia_Genes_Screen       1           N2     N2     N2    1.200000   \n",
      "1    Glia_Genes_Screen       1           N2     N2     N2    0.833333   \n",
      "2    Glia_Genes_Screen       1           N2     N2     N2    1.356667   \n",
      "3    Glia_Genes_Screen       1           N2     N2     N2    0.783333   \n",
      "4    Glia_Genes_Screen       1           N2     N2     N2    0.876667   \n",
      "..                 ...     ...          ...    ...    ...         ...   \n",
      "559  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452    0.330000   \n",
      "560  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452    0.853333   \n",
      "561  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452    0.566667   \n",
      "562  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452    0.560000   \n",
      "563  Glia_Genes_Screen       1  vab-3_ok452  vab-3  ok452    0.710000   \n",
      "\n",
      "     final_prob  final_speed  taps_y  habit_dura  habit_prob  habit_speed  \n",
      "0      0.047380     0.172764    29.0    1.470000    0.809763     0.025363  \n",
      "1      0.316363     0.252032    29.0    1.486667    0.628082    -0.025738  \n",
      "2      0.271435     0.240509    29.0    1.463333    0.585708     0.002753  \n",
      "3      0.291783     0.235854    29.0    1.946667    0.666551     0.019823  \n",
      "4      0.192593     0.244481    29.0    2.113333    0.682407     0.022408  \n",
      "..          ...          ...     ...         ...         ...          ...  \n",
      "559    0.304584     0.236088    29.0    2.360000    0.513598     0.046812  \n",
      "560    0.218081     0.239567    29.0    1.776667    0.698585     0.037999  \n",
      "561    0.300000     0.213481    29.0    2.313333    0.533333     0.054575  \n",
      "562    0.149539     0.217163    29.0    2.410000    0.692566     0.018864  \n",
      "563    0.206455     0.197815    29.0    2.210000    0.668545     0.076842  \n",
      "\n",
      "[551 rows x 18 columns]\n",
      "-------------------------------------------------------\n",
      "         time  recov_dura   dist  recov_prob  recov_speed  plate      Date  \\\n",
      "0    1189.999        2.04  0.505    0.405405     0.247549      1  20220919   \n",
      "1    1189.968        2.50  0.586    0.846154     0.234400      2  20220811   \n",
      "2    1189.985        1.17  0.319    0.538462     0.272650      3  20220811   \n",
      "3    1189.976        1.73  0.430    0.840000     0.248555      4  20220811   \n",
      "4    1189.959        3.52  1.004    0.700000     0.285227      5  20220811   \n",
      "..        ...         ...    ...         ...          ...    ...       ...   \n",
      "556  1189.965        1.59  0.537    0.916667     0.337736      5  20221105   \n",
      "557  1189.977        2.49  0.736    0.823529     0.295582      6  20221105   \n",
      "558  1189.975        1.82  0.544    0.933333     0.298901      7  20221105   \n",
      "559  1189.993        2.59  0.761    0.869565     0.293822      8  20221105   \n",
      "560  1189.977        2.65  0.804    0.774194     0.303396      9  20221105   \n",
      "\n",
      "    Plate_id             Screen  taps      dataset   Gene Allele  \n",
      "0      B1210  Glia_Genes_Screen    31           N2     N2     N2  \n",
      "1      C0811  Glia_Genes_Screen    31           N2     N2     N2  \n",
      "2      A0811  Glia_Genes_Screen    31           N2     N2     N2  \n",
      "3      C0811  Glia_Genes_Screen    31           N2     N2     N2  \n",
      "4      A0811  Glia_Genes_Screen    31           N2     N2     N2  \n",
      "..       ...                ...   ...          ...    ...    ...  \n",
      "556    A1105  Glia_Genes_Screen    31  vab-3_ok452  vab-3  ok452  \n",
      "557    B1105  Glia_Genes_Screen    31  vab-3_ok452  vab-3  ok452  \n",
      "558    C1105  Glia_Genes_Screen    31  vab-3_ok452  vab-3  ok452  \n",
      "559    C1105  Glia_Genes_Screen    31  vab-3_ok452  vab-3  ok452  \n",
      "560    B1105  Glia_Genes_Screen    31  vab-3_ok452  vab-3  ok452  \n",
      "\n",
      "[561 rows x 13 columns]\n",
      "-------------------------------------------------------\n",
      "-------------------------------------------------------\n",
      "     init_dura  init_prob  init_speed  plate      Date Plate_id  \\\n",
      "0         2.67   0.857143    0.198127      1  20220919    B1210   \n",
      "1         2.32   0.944444    0.226293      2  20220811    C0811   \n",
      "2         2.82   0.857143    0.243262      3  20220811    A0811   \n",
      "3         2.73   0.958333    0.255678      4  20220811    C0811   \n",
      "4         2.99   0.875000    0.266890      5  20220811    A0811   \n",
      "..         ...        ...         ...    ...       ...      ...   \n",
      "546       2.69   0.818182    0.282900      5  20221105    A1105   \n",
      "547       2.63   0.916667    0.277567      6  20221105    B1105   \n",
      "548       2.88   0.833333    0.268056      7  20221105    C1105   \n",
      "549       2.97   0.842105    0.236027      8  20221105    C1105   \n",
      "550       2.92   0.875000    0.274658      9  20221105    B1105   \n",
      "\n",
      "                Screen      dataset   Gene Allele  ...  habit_speed      time  \\\n",
      "0    Glia_Genes_Screen           N2     N2     N2  ...     0.025363  1189.999   \n",
      "1    Glia_Genes_Screen           N2     N2     N2  ...    -0.025738  1189.968   \n",
      "2    Glia_Genes_Screen           N2     N2     N2  ...     0.002753  1189.985   \n",
      "3    Glia_Genes_Screen           N2     N2     N2  ...     0.019823  1189.976   \n",
      "4    Glia_Genes_Screen           N2     N2     N2  ...     0.022408  1189.959   \n",
      "..                 ...          ...    ...    ...  ...          ...       ...   \n",
      "546  Glia_Genes_Screen  vab-3_ok452  vab-3  ok452  ...     0.046812  1189.965   \n",
      "547  Glia_Genes_Screen  vab-3_ok452  vab-3  ok452  ...     0.037999  1189.977   \n",
      "548  Glia_Genes_Screen  vab-3_ok452  vab-3  ok452  ...     0.054575  1189.975   \n",
      "549  Glia_Genes_Screen  vab-3_ok452  vab-3  ok452  ...     0.018864  1189.993   \n",
      "550  Glia_Genes_Screen  vab-3_ok452  vab-3  ok452  ...     0.076842  1189.977   \n",
      "\n",
      "     recov_dura   dist  recov_prob  recov_speed  taps  recovery_dura  \\\n",
      "0          2.04  0.505    0.405405     0.247549  31.0           0.63   \n",
      "1          2.50  0.586    0.846154     0.234400  31.0          -0.18   \n",
      "2          1.17  0.319    0.538462     0.272650  31.0           1.65   \n",
      "3          1.73  0.430    0.840000     0.248555  31.0           1.00   \n",
      "4          3.52  1.004    0.700000     0.285227  31.0          -0.53   \n",
      "..          ...    ...         ...          ...   ...            ...   \n",
      "546        1.59  0.537    0.916667     0.337736  31.0           1.10   \n",
      "547        2.49  0.736    0.823529     0.295582  31.0           0.14   \n",
      "548        1.82  0.544    0.933333     0.298901  31.0           1.06   \n",
      "549        2.59  0.761    0.869565     0.293822  31.0           0.38   \n",
      "550        2.65  0.804    0.774194     0.303396  31.0           0.27   \n",
      "\n",
      "     recovery_prob  recovery_speed  \n",
      "0         0.451737       -0.049422  \n",
      "1         0.098291       -0.008107  \n",
      "2         0.318681       -0.029387  \n",
      "3         0.118333        0.007123  \n",
      "4         0.175000       -0.018338  \n",
      "..             ...             ...  \n",
      "546      -0.098485       -0.054836  \n",
      "547       0.093137       -0.018016  \n",
      "548      -0.100000       -0.030846  \n",
      "549      -0.027460       -0.057795  \n",
      "550       0.100806       -0.028739  \n",
      "\n",
      "[542 rows x 25 columns]\n",
      "-------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "PD_tap=tap_output\n",
    "PD_first_tap = PD_tap[(PD_tap.taps==1)].reset_index().drop(columns=\"index\")\n",
    "PD_last_three_taps = PD_tap[((PD_tap.taps>=28)&(PD_tap.taps<=30))].reset_index().drop(columns=\"index\")\n",
    "PD_recov_taps = PD_tap[(PD_tap.taps==31)].reset_index().drop(columns=\"index\")\n",
    "\n",
    "print(PD_first_tap)\n",
    "print(\"-\"*55)\n",
    "print(PD_last_three_taps)\n",
    "print(\"-\"*55)\n",
    "# print(PD_recov_taps)\n",
    "\n",
    "print(\"-\"*55)\n",
    "PD_final_taps=PD_last_three_taps.groupby([\"dataset\", \"Date\",\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"plate\"]).mean().reset_index()\n",
    "# print(PD_final_taps)\n",
    "PD_final_taps=PD_final_taps.rename(columns={\"dura\": \"final_dura\", \"prob\": \"final_prob\", \"speed\": \"final_speed\"}, errors=\"raise\")\n",
    "\n",
    "\n",
    "print(\"PD_final_taps\")\n",
    "print(PD_final_taps)\n",
    "print(\"-\"*55)\n",
    "\n",
    "\n",
    "PD_first_tap=PD_first_tap.rename(columns={\"dura\": \"init_dura\", \"prob\": \"init_prob\", \"speed\": \"init_speed\"}, errors=\"raise\")\n",
    "# print(PD_final_taps)\n",
    "# print(PD_first_tap)\n",
    "PD_habit_levels = pd.merge(PD_first_tap, PD_final_taps, on =['dataset','plate',\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"Date\"], how ='left')\n",
    "PD_habit_levels = PD_habit_levels.dropna()\n",
    "\n",
    "print('PD_habit_levels')\n",
    "print(PD_habit_levels)\n",
    "print(\"-\"*55)\n",
    "# print(PD_habit_levels)\n",
    "PD_habit_levels['habit_dura']=PD_habit_levels.init_dura-PD_habit_levels.final_dura\n",
    "PD_habit_levels['habit_prob']=PD_habit_levels.init_prob-PD_habit_levels.final_prob\n",
    "PD_habit_levels['habit_speed']=PD_habit_levels.init_speed-PD_habit_levels.final_speed\n",
    "# PD_habit_levels=PD_habit_levels.drop(columns=['time_x','time_y','dist_x','dist_y', 'Allele','Gene'])\n",
    "PD_habit_levels=PD_habit_levels.drop(columns=['time_x','time_y','dist_x','dist_y'])\n",
    "\n",
    "print('final version PD_habit_levels')\n",
    "print(PD_habit_levels)\n",
    "print(\"-\"*55)\n",
    "\n",
    "PD_recov_taps=PD_recov_taps.rename(columns={\"dura\": \"recov_dura\", \"prob\": \"recov_prob\", \"speed\":\"recov_speed\"})\n",
    "print(PD_recov_taps)\n",
    "print(\"-\"*55)\n",
    "PD_habituation = pd.merge(PD_habit_levels, PD_recov_taps, on =['dataset','plate',\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"Date\"], how ='left')\n",
    "\n",
    "# PD_habituation = PD_habit_levels\n",
    "PD_habituation = PD_habituation.drop(columns=['taps_x','taps_y'])\n",
    "# PD_habituation = PD_habituation.drop(columns=['taps_x','taps_y'])\n",
    "PD_habituation = PD_habituation.dropna()\n",
    "# print(PD_habituation)\n",
    "# PD_habituation['recovery_dura']=0\n",
    "# PD_habituation['recovery_prob']=0\n",
    "# PD_habituation['recovery_speed']=0\n",
    "# PD_habituation['dist']=0\n",
    "# PD_habituation['recov_prob']=0\n",
    "# PD_habituation['recov_dura']=0\n",
    "# PD_habituation['recov_speed']=0\n",
    "PD_habituation['recovery_dura']=PD_habituation.init_dura-PD_habituation.recov_dura\n",
    "PD_habituation['recovery_prob']=PD_habituation.init_prob-PD_habituation.recov_prob\n",
    "PD_habituation['recovery_speed']=PD_habituation.init_speed-PD_habituation.recov_speed\n",
    "tap_data=PD_habituation.dropna() \n",
    "tap_data=PD_habituation\n",
    "print(\"-\"*55)\n",
    "print(tap_data)\n",
    "# print(PD_recov_taps)\n",
    "# print(PD_first_tap)\n",
    "# print(PD_last_three_taps)\n",
    "print(\"-\"*55)\n",
    "# print(PD_habit_levels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc7a8074-dc26-47c1-9117-32b5ec1ab227",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### The below cell reads in file from online repository (github) - example in case this is possible in the future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33cd988a-0fe1-48b3-9296-ec3d24c04f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # tap_url = 'https://osf.io/du9bj/files/osfstorage/650a2f9f1e76a4230e8a99a5?raw=true'\n",
    "# tap_url='https://github.com/MyYummyPancake/NRSC510B/blob/main/tap_output.csv?raw=true'\n",
    "# # s=requests.get(tap_url).content\n",
    "# # tap_output=pd.read_csv(io.StringIO(s.decode('utf-8')))\n",
    "# tap_output=pd.read_csv(tap_url, on_bad_lines='skip', index_col=0)\n",
    "# print(tap_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "762927ac-9ce8-444b-aa5b-3ebda2c8e889",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3e2bfea-63a1-4a26-8797-14757f0a2a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(tap_output['Gene'].unique())\n",
    "# print(len(tap_output['Gene'].unique()))\n",
    "# print(baseline_output['Gene'].unique())\n",
    "# print(len(baseline_output['Gene'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c617f17-02de-41fd-ae3d-3ab7565454b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for x in tap_output['Gene'].unique():\n",
    "#     tap_output_gene=tap_output[tap_output['Gene']==x]\n",
    "#     gene_tap_data=tap_output[tap_output['Date'].isin(tap_output_gene['Date'].unique())]\n",
    "#     gene_tap_data_final=gene_tap_data[gene_tap_data['Gene'].isin(['N2', x])]\n",
    "#     gene_tap_data_final['taps']=gene_tap_data_final['taps'].astype(int)\n",
    "\n",
    "\n",
    "# for x in baseline_output['Gene'].unique():\n",
    "#     baseline_output_gene=baseline_output[baseline_output['Gene']==x]\n",
    "#     gene_baseline_data=baseline_output[baseline_output['Date'].isin(baseline_output_gene['Date'].unique())]\n",
    "#     gene_baseline_data_final=gene_baseline_data[gene_baseline_data['Gene'].isin(['N2', x])]\n",
    "#     for a,b in zip(list_baseline_metrics, list_baseline_Tstats):\n",
    "#         TTest_Allele(x, a, baseline_output, b)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f49778f-28c4-42fa-8cec-a70569ef9c8c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03754e95-f034-4b76-bec0-38db8d2bdd00",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf79ef13-4d33-44dc-ac52-bdb776b5c5da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "74308e12-a414-4899-b18f-ed97d006930d",
   "metadata": {},
   "source": [
    "# Run Statistics (T-Test and sample-mean distance) on Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8603706-af3a-4596-bc65-320c51e086de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(baseline_output)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e867fef6-9b3c-4591-9755-534e26701b2d",
   "metadata": {},
   "source": [
    "## Sample Mean Distance & Confidence Interval Calculation of Baseline Data\n",
    "1. Calculate Mean\n",
    "\n",
    "2. Calculate SEM\n",
    "\n",
    "3. Calculate CI\n",
    "\n",
    "4. Calculate Mean Distance From Control (Mean of Strain - Mean of Control)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5eeba2f-d290-4781-8a36-ca31118a2889",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tap_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dd76fdc5-2356-4617-8c2c-29e1685808bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Gene  Instantaneous Speed  Interval Speed      Bias     Width  \\\n",
      "0        N2             0.071012        0.075835  0.150125  0.084687   \n",
      "1     ced-1             0.066342        0.070083  0.149105  0.088107   \n",
      "2     ced-1             0.042747        0.048264  0.055563  0.085415   \n",
      "3     ced-1             0.107864        0.086273  0.416429  0.080042   \n",
      "4     ced-1             0.115607        0.088923  0.514944  0.087389   \n",
      "..      ...                  ...             ...       ...       ...   \n",
      "268  lgc-34             0.069017        0.080340  0.205929  0.114775   \n",
      "269   vab-3             0.059317        0.066440  0.152330  0.104552   \n",
      "270      N2             0.077907        0.080118  0.312448  0.108200   \n",
      "271  lgc-34             0.068264        0.083968  0.212251  0.105617   \n",
      "272   vab-3             0.083227        0.083403  0.262274  0.095135   \n",
      "\n",
      "       Length      Area  Angular Speed  Aspect Ratio       Kink      Curve  \\\n",
      "0    1.138129  0.122129       3.948523      0.259845  45.110799  29.210423   \n",
      "1    1.144676  0.125379       3.544327      0.253217  44.130579  27.966616   \n",
      "2    1.070366  0.113615       2.826147      0.252087  42.621836  27.137727   \n",
      "3    0.974515  0.099616       5.772076      0.230054  40.123498  24.893404   \n",
      "4    1.091038  0.118326       4.969391      0.239320  40.519961  27.993482   \n",
      "..        ...       ...            ...           ...        ...        ...   \n",
      "268  1.072856  0.145503       5.026244      0.300133  51.690694  30.863091   \n",
      "269  1.008355  0.127625       3.657281      0.269276  48.319255  28.056455   \n",
      "270  1.143332  0.145596       3.579331      0.260219  46.184121  29.998864   \n",
      "271  1.065342  0.136100       4.061412      0.277342  49.171651  29.763572   \n",
      "272  0.953548  0.113127       7.315068      0.310163  51.046575  30.850906   \n",
      "\n",
      "         Crab  \n",
      "0    0.009792  \n",
      "1    0.008796  \n",
      "2    0.007080  \n",
      "3    0.010737  \n",
      "4    0.010665  \n",
      "..        ...  \n",
      "268  0.015254  \n",
      "269  0.012193  \n",
      "270  0.009999  \n",
      "271  0.011088  \n",
      "272  0.015034  \n",
      "\n",
      "[273 rows x 12 columns]\n",
      "------------------------------\n",
      "Index(['Instantaneous Speed', 'Interval Speed', 'Bias', 'Width', 'Length',\n",
      "       'Area', 'Angular Speed', 'Aspect Ratio', 'Kink', 'Curve', 'Crab'],\n",
      "      dtype='object')\n",
      "------------------------------\n",
      "       Gene  init_dura  init_prob  init_speed  final_dura  final_prob  \\\n",
      "0        N2      2.370   0.633711    0.266839    1.256667    0.193291   \n",
      "1     ced-1      2.145   0.535023    0.251951    0.756667    0.190239   \n",
      "2     ced-1      2.510   0.567073    0.200703    0.981667    0.302331   \n",
      "3     ced-1      3.265   0.812150    0.281833    1.030000    0.253394   \n",
      "4     ced-1      3.365   0.806620    0.279484    1.073333    0.179388   \n",
      "..      ...        ...        ...         ...         ...         ...   \n",
      "264  lgc-34      2.445   0.748264    0.270586    0.466667    0.032866   \n",
      "265   vab-3      2.925   0.837719    0.252041    0.563333    0.224769   \n",
      "266      N2      2.720   0.812500    0.255747    0.850000    0.294483   \n",
      "267  lgc-34      2.365   0.690476    0.256849    0.903333    0.060063   \n",
      "268   vab-3      2.590   0.750712    0.271589    0.900000    0.287638   \n",
      "\n",
      "     final_speed  habit_dura  habit_prob  habit_speed       time  taps  \\\n",
      "0       0.247360    1.113333    0.440419     0.019478  1189.9775  31.0   \n",
      "1       0.206591    1.388333    0.344784     0.045360  1189.9785  31.0   \n",
      "2       0.203948    1.528333    0.264742    -0.003244  1189.9805  31.0   \n",
      "3       0.223646    2.235000    0.558756     0.058188  1189.9760  31.0   \n",
      "4       0.245156    2.291667    0.627232     0.034328  1189.9765  31.0   \n",
      "..           ...         ...         ...          ...        ...   ...   \n",
      "264     0.296141    1.978333    0.715398    -0.025555  1189.9950  31.0   \n",
      "265     0.215322    2.361667    0.612950     0.036720  1189.9840  31.0   \n",
      "266     0.216973    1.870000    0.518017     0.038774  1189.9885  31.0   \n",
      "267     0.194217    1.461667    0.630413     0.062632  1189.9750  31.0   \n",
      "268     0.222423    1.690000    0.463075     0.049166  1189.9825  31.0   \n",
      "\n",
      "     recovery_dura  recovery_prob  recovery_speed  \n",
      "0           -1.415       0.254626       -0.007369  \n",
      "1            0.195      -0.092155        0.004845  \n",
      "2            0.675      -0.151742       -0.006987  \n",
      "3            0.275       0.183579        0.016366  \n",
      "4           -0.030       0.261472        0.003728  \n",
      "..             ...            ...             ...  \n",
      "264          0.520       0.335945       -0.020220  \n",
      "265          0.720      -0.063730       -0.044321  \n",
      "266          0.485      -0.079265        0.009047  \n",
      "267          0.950       0.426407        0.013554  \n",
      "268          0.450      -0.058316        0.013851  \n",
      "\n",
      "[269 rows x 15 columns]\n",
      "------------------------------\n",
      "Index(['Gene', 'init_dura', 'init_prob', 'init_speed', 'final_dura',\n",
      "       'final_prob', 'final_speed', 'habit_dura', 'habit_prob', 'habit_speed',\n",
      "       'time', 'taps', 'recovery_dura', 'recovery_prob', 'recovery_speed'],\n",
      "      dtype='object')\n",
      "--------------------------------------------------\n",
      "allele output\n",
      "--------------------------------------------------\n",
      "          dataset  Instantaneous Speed  Interval Speed      Bias     Width  \\\n",
      "0              N2             0.071012        0.075835  0.150125  0.084687   \n",
      "1     ced-1_e1735             0.066342        0.070083  0.149105  0.088107   \n",
      "2     ced-1_e1754             0.042747        0.048264  0.055563  0.085415   \n",
      "3     ced-1_n1995             0.107864        0.086273  0.416429  0.080042   \n",
      "4     ced-1_n2089             0.115607        0.088923  0.514944  0.087389   \n",
      "..            ...                  ...             ...       ...       ...   \n",
      "268  lgc-34_gk532             0.069017        0.080340  0.205929  0.114775   \n",
      "269   vab-3_ok452             0.059317        0.066440  0.152330  0.104552   \n",
      "270            N2             0.077907        0.080118  0.312448  0.108200   \n",
      "271  lgc-34_gk532             0.068264        0.083968  0.212251  0.105617   \n",
      "272   vab-3_ok452             0.083227        0.083403  0.262274  0.095135   \n",
      "\n",
      "       Length      Area  Angular Speed  Aspect Ratio       Kink      Curve  \\\n",
      "0    1.138129  0.122129       3.948523      0.259845  45.110799  29.210423   \n",
      "1    1.144676  0.125379       3.544327      0.253217  44.130579  27.966616   \n",
      "2    1.070366  0.113615       2.826147      0.252087  42.621836  27.137727   \n",
      "3    0.974515  0.099616       5.772076      0.230054  40.123498  24.893404   \n",
      "4    1.091038  0.118326       4.969391      0.239320  40.519961  27.993482   \n",
      "..        ...       ...            ...           ...        ...        ...   \n",
      "268  1.072856  0.145503       5.026244      0.300133  51.690694  30.863091   \n",
      "269  1.008355  0.127625       3.657281      0.269276  48.319255  28.056455   \n",
      "270  1.143332  0.145596       3.579331      0.260219  46.184121  29.998864   \n",
      "271  1.065342  0.136100       4.061412      0.277342  49.171651  29.763572   \n",
      "272  0.953548  0.113127       7.315068      0.310163  51.046575  30.850906   \n",
      "\n",
      "         Crab  \n",
      "0    0.009792  \n",
      "1    0.008796  \n",
      "2    0.007080  \n",
      "3    0.010737  \n",
      "4    0.010665  \n",
      "..        ...  \n",
      "268  0.015254  \n",
      "269  0.012193  \n",
      "270  0.009999  \n",
      "271  0.011088  \n",
      "272  0.015034  \n",
      "\n",
      "[273 rows x 12 columns]\n",
      "--------------------------------------------------\n",
      "Index(['dataset', 'Instantaneous Speed', 'Interval Speed', 'Bias', 'Width',\n",
      "       'Length', 'Area', 'Angular Speed', 'Aspect Ratio', 'Kink', 'Curve',\n",
      "       'Crab'],\n",
      "      dtype='object')\n",
      "--------------------------------------------------\n",
      "          dataset  init_dura  init_prob  init_speed  final_dura  final_prob  \\\n",
      "0              N2      2.370   0.633711    0.266839    1.256667    0.193291   \n",
      "1     ced-1_e1735      2.145   0.535023    0.251951    0.756667    0.190239   \n",
      "2     ced-1_e1754      2.510   0.567073    0.200703    0.981667    0.302331   \n",
      "3     ced-1_n1995      3.265   0.812150    0.281833    1.030000    0.253394   \n",
      "4     ced-1_n2089      3.365   0.806620    0.279484    1.073333    0.179388   \n",
      "..            ...        ...        ...         ...         ...         ...   \n",
      "264  lgc-34_gk532      2.445   0.748264    0.270586    0.466667    0.032866   \n",
      "265   vab-3_ok452      2.925   0.837719    0.252041    0.563333    0.224769   \n",
      "266            N2      2.720   0.812500    0.255747    0.850000    0.294483   \n",
      "267  lgc-34_gk532      2.365   0.690476    0.256849    0.903333    0.060063   \n",
      "268   vab-3_ok452      2.590   0.750712    0.271589    0.900000    0.287638   \n",
      "\n",
      "     final_speed  habit_dura  habit_prob  habit_speed       time  taps  \\\n",
      "0       0.247360    1.113333    0.440419     0.019478  1189.9775  31.0   \n",
      "1       0.206591    1.388333    0.344784     0.045360  1189.9785  31.0   \n",
      "2       0.203948    1.528333    0.264742    -0.003244  1189.9805  31.0   \n",
      "3       0.223646    2.235000    0.558756     0.058188  1189.9760  31.0   \n",
      "4       0.245156    2.291667    0.627232     0.034328  1189.9765  31.0   \n",
      "..           ...         ...         ...          ...        ...   ...   \n",
      "264     0.296141    1.978333    0.715398    -0.025555  1189.9950  31.0   \n",
      "265     0.215322    2.361667    0.612950     0.036720  1189.9840  31.0   \n",
      "266     0.216973    1.870000    0.518017     0.038774  1189.9885  31.0   \n",
      "267     0.194217    1.461667    0.630413     0.062632  1189.9750  31.0   \n",
      "268     0.222423    1.690000    0.463075     0.049166  1189.9825  31.0   \n",
      "\n",
      "     recovery_dura  recovery_prob  recovery_speed  \n",
      "0           -1.415       0.254626       -0.007369  \n",
      "1            0.195      -0.092155        0.004845  \n",
      "2            0.675      -0.151742       -0.006987  \n",
      "3            0.275       0.183579        0.016366  \n",
      "4           -0.030       0.261472        0.003728  \n",
      "..             ...            ...             ...  \n",
      "264          0.520       0.335945       -0.020220  \n",
      "265          0.720      -0.063730       -0.044321  \n",
      "266          0.485      -0.079265        0.009047  \n",
      "267          0.950       0.426407        0.013554  \n",
      "268          0.450      -0.058316        0.013851  \n",
      "\n",
      "[269 rows x 15 columns]\n",
      "--------------------------------------------------\n",
      "Index(['dataset', 'init_dura', 'init_prob', 'init_speed', 'final_dura',\n",
      "       'final_prob', 'final_speed', 'habit_dura', 'habit_prob', 'habit_speed',\n",
      "       'time', 'taps', 'recovery_dura', 'recovery_prob', 'recovery_speed'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "baseline_output_byplate=baseline_output.groupby(by=['Plate_id',\n",
    "                                                    'Date',\n",
    "                                                    'Screen',\n",
    "                                                    'dataset',\n",
    "                                                    'Gene',\n",
    "                                                    'Allele'],as_index=False).mean().drop(columns=['Plate_id',\n",
    "                                                                                                    'n',\n",
    "                                                                                                    'Number',\n",
    "                                                                                                   'Date',\n",
    "                                                                                                    'Time',\n",
    "                                                                                                    'Screen',\n",
    "                                                                                                    'dataset',\n",
    "                                                                                                    'Allele'])\n",
    "print(baseline_output_byplate)\n",
    "print('-'*30)\n",
    "print(baseline_output_byplate.columns[1:])\n",
    "# print(baseline_output_byplate[['Gene','Instantaneous Speed']].copy())\n",
    "\n",
    "print('-'*30)\n",
    "tap_data_byplate=tap_data.groupby(by=['Plate_id',\n",
    "                                    'Date',\n",
    "                                    'Screen',\n",
    "                                    'dataset',\n",
    "                                    'Gene',\n",
    "                                    'Allele'],as_index=False).mean().drop(columns=['Plate_id',\n",
    "                                                                                    'Screen',\n",
    "                                                                                    'dataset',\n",
    "                                                                                    'Allele',\n",
    "                                                                                   'Date',\n",
    "                                                                                    'dist',\n",
    "                                                                                    'plate',\n",
    "                                                                                    'recov_dura',\n",
    "                                                                                    'recov_prob',\n",
    "                                                                                    'recov_speed'])\n",
    "\n",
    "\n",
    "print(tap_data_byplate)\n",
    "print('-'*30)\n",
    "print(tap_data_byplate.columns)\n",
    "print('-'*50)\n",
    "\n",
    "print(\"allele output\")\n",
    "print('-'*50)\n",
    "baseline_output_allele_byplate=baseline_output.groupby(by=['Plate_id',\n",
    "                                                            'Date',\n",
    "                                                            'Screen',\n",
    "                                                            'dataset',\n",
    "                                                            'Gene',\n",
    "                                                            'Allele'],as_index=False).mean().drop(columns=['Plate_id',\n",
    "                                                                                                            'n',\n",
    "                                                                                                            'Number',\n",
    "                                                                                                            'Time',\n",
    "                                                                                                            'Screen',\n",
    "                                                                                                           'Date',\n",
    "                                                                                                            'Gene',\n",
    "                                                                                                            'Allele'])\n",
    "print(baseline_output_allele_byplate)\n",
    "print('-'*50)\n",
    "print(baseline_output_allele_byplate.columns)\n",
    "# print(baseline_output_byplate[['Gene','Instantaneous Speed']].copy())\n",
    "\n",
    "print('-'*50)\n",
    "tap_data_allele_byplate=tap_data.groupby(by=['Plate_id',\n",
    "                                            'Date',\n",
    "                                            'Screen',\n",
    "                                            'dataset',\n",
    "                                            'Gene',\n",
    "                                            'Allele'],as_index=False).mean().drop(columns=['Plate_id',\n",
    "                                                                                            'Screen',\n",
    "                                                                                            'Gene',\n",
    "                                                                                           'Date',\n",
    "                                                                                            'Allele',\n",
    "                                                                                            'dist',\n",
    "                                                                                            'plate',\n",
    "                                                                                            'recov_dura',\n",
    "                                                                                            'recov_prob',\n",
    "                                                                                            'recov_speed'])\n",
    "\n",
    "\n",
    "print(tap_data_allele_byplate)\n",
    "print('-'*50)\n",
    "print(tap_data_allele_byplate.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8304d782-9424-4068-a3f4-4982b1be21fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(tap_data_allele_byplate[tap_data_allele_byplate.dataset=='N2_XJ1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "10502cca-0ab9-4bd5-9801-7a247a6af822",
   "metadata": {},
   "outputs": [],
   "source": [
    "''' \n",
    "input: one pd.DataFrame with multiple columns (1st column is the index, the other are phenotypes)\n",
    "return: list of pd.DataFrame, each DataFrame with 2 columns: index and 1 phenotype, \n",
    "        for how many phenotypes there are in the input\n",
    "'''\n",
    "def extract_phenotypes(df):\n",
    "    list_phenotypes_df = []\n",
    "    phenotypes = df.columns[1:]\n",
    "    index = df.columns[0]\n",
    "    for i in phenotypes:\n",
    "        new_df = df[[index, i]].copy()\n",
    "        list_phenotypes_df.append(new_df)\n",
    "    return list_phenotypes_df\n",
    "\n",
    "\n",
    "''' \n",
    "input: one pd.DataFrame with multiple columns (1st column is the index, the other are phenotypes)\n",
    "return: list of pd.DataFrame, each DataFrame with 2 columns: index and 1 phenotype, \n",
    "        for how many phenotypes there are in the input\n",
    "'''\n",
    "def extract_allele_phenotypes(df):\n",
    "    list_phenotypes_df = []\n",
    "    phenotypes = df.columns[1:]\n",
    "    index = df.columns[0]\n",
    "    for i in phenotypes:\n",
    "        new_df = df[[index, i]].copy()\n",
    "        list_phenotypes_df.append(new_df)\n",
    "    return list_phenotypes_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "input: list of dataframes with 2 columns (1st one is index, 2nd one are indiviudal phenotypes)\n",
    "        output from extract_phenotypes function\n",
    "\n",
    "output: the same list of dataframes with additional calculated columns for:\n",
    "                1. mean\n",
    "                2. count\n",
    "                3. std\n",
    "                4. ci95_hi\n",
    "                5. ci95_low\n",
    "                6. ci95\n",
    "'''\n",
    "        # baseline_output_gene=baseline_output[baseline_output['dataset']==x]\n",
    "        # gene_baseline_data=baseline_output[baseline_output['Date'].isin(baseline_output_gene['Date'].unique())]\n",
    "        # gene_baseline_data_final=gene_baseline_data[gene_baseline_data['dataset'].isin(['N2', x])]\n",
    "\n",
    "def calculate_MSD(list_of_dfs):\n",
    "    new_list_of_dfs=[]\n",
    "    for i in list_of_dfs:\n",
    "        calculate_1= i.groupby(by=['Gene'], as_index=False).agg(['mean', 'count', 'sem'])\n",
    "        calculate_2= ci95(calculate_1)\n",
    "        N2_data = calculate_2[calculate_2.index=='N2']\n",
    "        calculate_2.iloc[:,0] = calculate_2.iloc[:,0]-N2_data.iloc[0,0]\n",
    "        calculate_2.iloc[:,3] = calculate_2.iloc[:,3]-N2_data.iloc[0,0]\n",
    "        calculate_2.iloc[:,4] = calculate_2.iloc[:,4]-N2_data.iloc[0,0]\n",
    "      \n",
    "        \n",
    "        # calculate_2.columns=['_'.join(col) for col in calculate_2.columns.values]\n",
    "        new_list_of_dfs.append(calculate_2)\n",
    "    return new_list_of_dfs\n",
    "    \n",
    "\n",
    "\n",
    "'''\n",
    "input: df of 4 columns: index, mean, count, std\n",
    "\n",
    "output: df of 6 columns: index, mean, count, std, ci95_hi, ci95_low\n",
    "\n",
    "'''\n",
    "def ci95(df):\n",
    "    for metric in df.columns.levels[0]:\n",
    "        if metric == 'Gene':\n",
    "            pass\n",
    "        else:\n",
    "            ci95_hi = []\n",
    "            ci95_lo = []\n",
    "            for i in df[metric].index:\n",
    "                m, c, s = df[metric].loc[i]\n",
    "                ci95_hi.append(stats.t.interval(confidence=0.95, df=c-1, loc=m, scale=s)[1])\n",
    "                ci95_lo.append(stats.t.interval(confidence=0.95, df=c-1, loc=m, scale=s)[0])\n",
    "            df[metric,'ci95_hi'] = ci95_hi\n",
    "            df[metric,'ci95_lo'] = ci95_lo\n",
    "            # df[metric,'ci95']=list(zip(ci95_lo,ci95_hi))\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "input: list of dataframes with 2 columns (1st one is index, 2nd one are indiviudal phenotypes)\n",
    "        output from extract_phenotypes function\n",
    "\n",
    "output: the same list of dataframes with additional calculated columns for:\n",
    "                1. mean\n",
    "                2. count\n",
    "                3. std\n",
    "                4. ci95_hi\n",
    "                5. ci95_low\n",
    "                6. ci95\n",
    "'''\n",
    "        # baseline_output_gene=baseline_output[baseline_output['dataset']==x]\n",
    "        # gene_baseline_data=baseline_output[baseline_output['Date'].isin(baseline_output_gene['Date'].unique())]\n",
    "        # gene_baseline_data_final=gene_baseline_data[gene_baseline_data['dataset'].isin(['N2', x])]\n",
    "\n",
    "def calculate_allele_MSD(list_of_dfs):\n",
    "    new_list_of_dfs=[]\n",
    "    for i in list_of_dfs:\n",
    "        calculate_1= i.groupby(by=['dataset'], as_index=False).agg(['mean', 'count', 'sem'])\n",
    "        calculate_2= ci95_allele(calculate_1)\n",
    "        N2_data = calculate_2[calculate_2.index=='N2_XJ1']\n",
    "        calculate_2.iloc[:,0] = calculate_2.iloc[:,0]-N2_data.iloc[0,0]\n",
    "        calculate_2.iloc[:,3] = calculate_2.iloc[:,3]-N2_data.iloc[0,0]\n",
    "        calculate_2.iloc[:,4] = calculate_2.iloc[:,4]-N2_data.iloc[0,0]\n",
    "        # calculate_2.columns=['_'.join(col) for col in calculate_2.columns.values]\n",
    "        new_list_of_dfs.append(calculate_2)\n",
    "    return new_list_of_dfs\n",
    "    \n",
    "        \n",
    "'''\n",
    "input: df of 4 columns: index, mean, count, std\n",
    "\n",
    "output: df of 6 columns: index, mean, count, std, ci95_hi, ci95_low\n",
    "\n",
    "'''\n",
    "def ci95_allele(df):\n",
    "    for metric in df.columns.levels[0]:\n",
    "        if metric == 'dataset':\n",
    "            pass\n",
    "        else:\n",
    "            ci95_hi = []\n",
    "            ci95_lo = []\n",
    "            for i in df[metric].index:\n",
    "                m, c, s = df[metric].loc[i]\n",
    "                ci95_hi.append(stats.t.interval(confidence=0.95, df=c-1, loc=m, scale=s)[1])\n",
    "                ci95_lo.append(stats.t.interval(confidence=0.95, df=c-1, loc=m, scale=s)[0])\n",
    "            df[metric,'ci95_hi'] = ci95_hi\n",
    "            df[metric,'ci95_lo'] = ci95_lo\n",
    "            # df[metric,'ci95']=list(zip(ci95_lo,ci95_hi))\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1fc7e03c-de00-436d-b9f0-fcaf1063188b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[                    Instantaneous Speed                                    \n",
      "                                   mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                       \n",
      "N2                             0.000000    57  0.004099  0.008212 -0.008212\n",
      "acd-1xdelm-2                   0.034337     5  0.012477  0.068977 -0.000303\n",
      "acd-1xdelm-2xdelm-1            0.036644     5  0.010984  0.067141  0.006148\n",
      "alr-1                         -0.013400    18  0.001972 -0.009239 -0.017561\n",
      "ced-1                          0.002253    30  0.004833  0.012138 -0.007631\n",
      "ced-12                        -0.018417    12  0.004997 -0.007420 -0.029415\n",
      "cnd-1                         -0.039882    18  0.002180 -0.035283 -0.044482\n",
      "dyf-11                        -0.033216    30  0.002029 -0.029066 -0.037367\n",
      "gcy-18                         0.005938     4  0.008678  0.033556 -0.021680\n",
      "gcy-23                        -0.028251     9  0.002755 -0.021897 -0.034604\n",
      "gcy-8                         -0.022238     9  0.003740 -0.013614 -0.030863\n",
      "lgc-34                        -0.003764     5  0.003690  0.006482 -0.014010\n",
      "mls-2                         -0.028669    12  0.002577 -0.022996 -0.034342\n",
      "ngn-1                         -0.049634     3  0.001582 -0.042828 -0.056440\n",
      "pde-1                          0.010720     9  0.007165  0.027242 -0.005802\n",
      "psr-1                         -0.014762     6  0.005871  0.000330 -0.029855\n",
      "scrm-1                        -0.016341     6  0.008475  0.005444 -0.038126\n",
      "tax-2                          0.000973    24  0.008052  0.017631 -0.015684\n",
      "ttx-1                         -0.024291     6  0.002758 -0.017201 -0.031381\n",
      "vab-3                         -0.006715     5  0.005972  0.009865 -0.023295,                     Interval Speed                                    \n",
      "                              mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                  \n",
      "N2                        0.000000    57  0.002109  0.004225 -0.004225\n",
      "acd-1xdelm-2              0.017776     5  0.005104  0.031948  0.003604\n",
      "acd-1xdelm-2xdelm-1       0.025274     5  0.004213  0.036970  0.013578\n",
      "alr-1                    -0.009968    18  0.002023 -0.005700 -0.014235\n",
      "ced-1                    -0.005148    30  0.003135  0.001264 -0.011560\n",
      "ced-12                   -0.016866    12  0.005847 -0.003998 -0.029735\n",
      "cnd-1                    -0.050563    18  0.001331 -0.047756 -0.053371\n",
      "dyf-11                   -0.031229    30  0.002837 -0.025426 -0.037032\n",
      "gcy-18                    0.014342     4  0.006961  0.036495 -0.007811\n",
      "gcy-23                   -0.021100     9  0.003741 -0.012474 -0.029726\n",
      "gcy-8                    -0.012791     9  0.005330 -0.000499 -0.025083\n",
      "lgc-34                    0.007405     5  0.005197  0.021834 -0.007024\n",
      "mls-2                    -0.024855    12  0.003660 -0.016801 -0.032910\n",
      "ngn-1                    -0.060042     3  0.003514 -0.044924 -0.075161\n",
      "pde-1                     0.002739     9  0.003427  0.010642 -0.005164\n",
      "psr-1                    -0.006011     6  0.007842  0.014147 -0.026169\n",
      "scrm-1                   -0.015075     6  0.007272  0.003619 -0.033769\n",
      "tax-2                    -0.006320    24  0.004155  0.002275 -0.014915\n",
      "ttx-1                    -0.017697     6  0.004050 -0.007285 -0.028108\n",
      "vab-3                    -0.002140     5  0.004496  0.010345 -0.014624,                          Bias                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.018201  0.036461 -0.036461\n",
      "acd-1xdelm-2         0.187328     5  0.080986  0.412180 -0.037525\n",
      "acd-1xdelm-2xdelm-1  0.185885     5  0.062675  0.359899  0.011871\n",
      "alr-1               -0.076979    18  0.010798 -0.054197 -0.099761\n",
      "ced-1                0.064212    30  0.031600  0.128842 -0.000418\n",
      "ced-12              -0.086606    12  0.027529 -0.026016 -0.147197\n",
      "cnd-1               -0.178350    18  0.011478 -0.154132 -0.202567\n",
      "dyf-11              -0.161973    30  0.010801 -0.139882 -0.184064\n",
      "gcy-18               0.050496     4  0.029265  0.143632 -0.042639\n",
      "gcy-23              -0.144505     9  0.018948 -0.100811 -0.188199\n",
      "gcy-8               -0.096357     9  0.020193 -0.049791 -0.142923\n",
      "lgc-34              -0.002062     5  0.029831  0.080761 -0.084886\n",
      "mls-2               -0.100747    12  0.013824 -0.070321 -0.131172\n",
      "ngn-1               -0.232935     3  0.003376 -0.218410 -0.247459\n",
      "pde-1                0.117759     9  0.049907  0.232845  0.002673\n",
      "psr-1               -0.054744     6  0.040979  0.050597 -0.160085\n",
      "scrm-1              -0.081625     6  0.034613  0.007351 -0.170601\n",
      "tax-2                0.030936    24  0.042169  0.118169 -0.056296\n",
      "ttx-1               -0.116657     6  0.014735 -0.078780 -0.154535\n",
      "vab-3               -0.048413     5  0.032201  0.040992 -0.137817,                         Width                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.001573  0.003151 -0.003151\n",
      "acd-1xdelm-2        -0.009956     5  0.002881 -0.001958 -0.017953\n",
      "acd-1xdelm-2xdelm-1 -0.008981     5  0.004282  0.002908 -0.020870\n",
      "alr-1               -0.012200    18  0.003425 -0.004975 -0.019425\n",
      "ced-1               -0.011711    30  0.001743 -0.008146 -0.015276\n",
      "ced-12              -0.005010    12  0.003284  0.002218 -0.012239\n",
      "cnd-1               -0.011885    18  0.001997 -0.007672 -0.016099\n",
      "dyf-11              -0.021456    30  0.001555 -0.018275 -0.024637\n",
      "gcy-18              -0.001651     4  0.002483  0.006252 -0.009555\n",
      "gcy-23              -0.011976     9  0.001734 -0.007979 -0.015974\n",
      "gcy-8                0.004672     9  0.007517  0.022005 -0.012662\n",
      "lgc-34               0.000642     5  0.003753  0.011061 -0.009778\n",
      "mls-2               -0.014417    12  0.002841 -0.008165 -0.020670\n",
      "ngn-1               -0.008769     3  0.007830  0.024919 -0.042457\n",
      "pde-1               -0.006700     9  0.003406  0.001155 -0.014555\n",
      "psr-1               -0.003376     6  0.003138  0.004691 -0.011443\n",
      "scrm-1               0.001585     6  0.003346  0.010186 -0.007016\n",
      "tax-2                0.002705    24  0.003814  0.010594 -0.005185\n",
      "ttx-1               -0.016441     6  0.001194 -0.013372 -0.019510\n",
      "vab-3               -0.010897     5  0.002414 -0.004194 -0.017601,                        Length                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.007447  0.014919 -0.014919\n",
      "acd-1xdelm-2        -0.018698     5  0.039150  0.089999 -0.127394\n",
      "acd-1xdelm-2xdelm-1 -0.023961     5  0.016892  0.022938 -0.070861\n",
      "alr-1               -0.220099    18  0.021162 -0.175452 -0.264746\n",
      "ced-1               -0.076827    30  0.012659 -0.050938 -0.102717\n",
      "ced-12              -0.130080    12  0.018318 -0.089763 -0.170398\n",
      "cnd-1               -0.274084    18  0.013514 -0.245572 -0.302597\n",
      "dyf-11              -0.237993    30  0.014374 -0.208595 -0.267390\n",
      "gcy-18              -0.003452     4  0.028546  0.087393 -0.094297\n",
      "gcy-23              -0.100898     9  0.022719 -0.048507 -0.153289\n",
      "gcy-8               -0.060358     9  0.014338 -0.027294 -0.093421\n",
      "lgc-34              -0.066751     5  0.009544 -0.040253 -0.093249\n",
      "mls-2               -0.126401    12  0.016740 -0.089556 -0.163246\n",
      "ngn-1               -0.287399     3  0.021095 -0.196636 -0.378162\n",
      "pde-1               -0.040366     9  0.015463 -0.004709 -0.076023\n",
      "psr-1               -0.059570     6  0.010173 -0.033421 -0.085720\n",
      "scrm-1              -0.011944     6  0.030392  0.066181 -0.090068\n",
      "tax-2               -0.130434    24  0.014060 -0.101347 -0.159520\n",
      "ttx-1               -0.224707     6  0.014157 -0.188315 -0.261099\n",
      "vab-3               -0.156359     5  0.028855 -0.076246 -0.236472,                          Area                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.001979  0.003964 -0.003964\n",
      "acd-1xdelm-2        -0.008791     5  0.003442  0.000765 -0.018348\n",
      "acd-1xdelm-2xdelm-1 -0.011060     5  0.004755  0.002143 -0.024263\n",
      "alr-1               -0.035437    18  0.003731 -0.027565 -0.043310\n",
      "ced-1               -0.019651    30  0.002416 -0.014710 -0.024593\n",
      "ced-12              -0.020110    12  0.004265 -0.010723 -0.029496\n",
      "cnd-1               -0.043362    18  0.002370 -0.038362 -0.048361\n",
      "dyf-11              -0.045571    30  0.002474 -0.040510 -0.050631\n",
      "gcy-18              -0.002598     4  0.002143  0.004221 -0.009418\n",
      "gcy-23              -0.022394     9  0.002998 -0.015481 -0.029307\n",
      "gcy-8               -0.001151     9  0.008345  0.018093 -0.020394\n",
      "lgc-34              -0.006041     5  0.003606  0.003971 -0.016053\n",
      "mls-2               -0.027807    12  0.003662 -0.019747 -0.035867\n",
      "ngn-1               -0.043258     3  0.002026 -0.034540 -0.051976\n",
      "pde-1               -0.011330     9  0.003607 -0.003013 -0.019648\n",
      "psr-1               -0.009488     6  0.002077 -0.004149 -0.014826\n",
      "scrm-1               0.001270     6  0.004780  0.013556 -0.011016\n",
      "tax-2               -0.016170    24  0.002942 -0.010085 -0.022256\n",
      "ttx-1               -0.039875     6  0.001143 -0.036938 -0.042813\n",
      "vab-3               -0.027507     5  0.004661 -0.014565 -0.040448,                     Angular Speed                                    \n",
      "                             mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                 \n",
      "N2                       0.000000    57  0.200576  0.401802 -0.401802\n",
      "acd-1xdelm-2             2.234596     5  0.899593  4.732266 -0.263075\n",
      "acd-1xdelm-2xdelm-1      2.681390     5  0.804646  4.915445  0.447335\n",
      "alr-1                    0.648670    18  0.253083  1.182628  0.114712\n",
      "ced-1                   -0.383598    30  0.219635  0.065606 -0.832801\n",
      "ced-12                  -0.089521    12  0.340821  0.660621 -0.839663\n",
      "cnd-1                   -0.638352    18  0.178163 -0.262461 -1.014243\n",
      "dyf-11                  -1.058622    30  0.141866 -0.768473 -1.348770\n",
      "gcy-18                  -0.043059     4  0.371310  1.138616 -1.224733\n",
      "gcy-23                  -1.342754     9  0.131453 -1.039622 -1.645886\n",
      "gcy-8                   -0.830472     9  0.318166 -0.096779 -1.564164\n",
      "lgc-34                   0.719153     5  0.515094  2.149283 -0.710976\n",
      "mls-2                   -1.164842    12  0.204689 -0.714324 -1.615360\n",
      "ngn-1                   -0.707482     3  0.620746  1.963372 -3.378335\n",
      "pde-1                    0.237408     9  0.230640  0.769264 -0.294448\n",
      "psr-1                   -0.898011     6  0.231327 -0.303365 -1.492657\n",
      "scrm-1                  -1.002279     6  0.404033  0.036322 -2.040880\n",
      "tax-2                    2.328270    24  0.502058  3.366856  1.289683\n",
      "ttx-1                   -0.841897     6  0.106198 -0.568906 -1.114887\n",
      "vab-3                    1.478808     5  1.163808  4.710058 -1.752441,                     Aspect Ratio                                    \n",
      "                            mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                \n",
      "N2                      0.000000    57  0.002585  0.005179 -0.005179\n",
      "acd-1xdelm-2           -0.002171     5  0.003389  0.007237 -0.011580\n",
      "acd-1xdelm-2xdelm-1     0.002833     5  0.006069  0.019683 -0.014016\n",
      "alr-1                  -0.000615    18  0.006545  0.013194 -0.014425\n",
      "ced-1                  -0.031358    30  0.003142 -0.024932 -0.037784\n",
      "ced-12                  0.006192    12  0.006602  0.020723 -0.008340\n",
      "cnd-1                   0.091840    18  0.006999  0.106606  0.077075\n",
      "dyf-11                 -0.014902    30  0.004523 -0.005652 -0.024152\n",
      "gcy-18                 -0.004862     4  0.009309  0.024764 -0.034487\n",
      "gcy-23                 -0.014059     9  0.002564 -0.008147 -0.019971\n",
      "gcy-8                   0.008600     9  0.008232  0.027582 -0.010382\n",
      "lgc-34                  0.007590     5  0.005516  0.022906 -0.007725\n",
      "mls-2                  -0.007295    12  0.007407  0.009008 -0.023598\n",
      "ngn-1                   0.160591     3  0.083594  0.520266 -0.199084\n",
      "pde-1                  -0.014091     9  0.003259 -0.006576 -0.021606\n",
      "psr-1                  -0.008109     6  0.003299  0.000371 -0.016590\n",
      "scrm-1                 -0.014026     6  0.005060 -0.001019 -0.027033\n",
      "tax-2                   0.047704    24  0.003751  0.055464  0.039945\n",
      "ttx-1                  -0.006097     6  0.003717  0.003457 -0.015651\n",
      "vab-3                   0.016279     5  0.013175  0.052858 -0.020300,                           Kink                                       \n",
      "                          mean count        sem    ci95_hi    ci95_lo\n",
      "Gene                                                                 \n",
      "N2                    0.000000    57   0.476721   0.954987  -0.954987\n",
      "acd-1xdelm-2         -1.496497     5   1.210764   1.865124  -4.858118\n",
      "acd-1xdelm-2xdelm-1  -0.699168     5   0.824352   1.589598  -2.987935\n",
      "alr-1                -1.658392    18   0.866323   0.169389  -3.486173\n",
      "ced-1                -5.489927    30   0.545099  -4.375074  -6.604780\n",
      "ced-12                0.253741    12   0.795448   2.004511  -1.497029\n",
      "cnd-1                19.195986    18   1.389624  22.127838  16.264135\n",
      "dyf-11               -2.850258    30   0.648183  -1.524576  -4.175940\n",
      "gcy-18               -1.039681     4   1.163117   2.661875  -4.741237\n",
      "gcy-23               -2.166522     9   0.535435  -0.931807  -3.401236\n",
      "gcy-8                 0.838433     9   0.706762   2.468228  -0.791362\n",
      "lgc-34                2.613964     5   0.579681   4.223418   1.004511\n",
      "mls-2                 1.935932    12   1.594450   5.445293  -1.573428\n",
      "ngn-1                30.073054     3  13.707654  89.052331 -28.906223\n",
      "pde-1                -2.926271     9   0.860648  -0.941613  -4.910930\n",
      "psr-1                -2.219610     6   0.552758  -0.798701  -3.640519\n",
      "scrm-1               -0.930607     6   1.031248   1.720301  -3.581515\n",
      "tax-2                 6.445200    24   0.801732   8.103710   4.786690\n",
      "ttx-1                -1.636833     6   0.773134   0.350572  -3.624237\n",
      "vab-3                 1.743059     5   1.195596   5.062567  -1.576449,                         Curve                                      \n",
      "                         mean count       sem    ci95_hi    ci95_lo\n",
      "Gene                                                               \n",
      "N2                   0.000000    57  0.201007   0.402665  -0.402665\n",
      "acd-1xdelm-2        -0.695415     5  0.721686   1.308306  -2.699137\n",
      "acd-1xdelm-2xdelm-1 -0.831539     5  0.426569   0.352807  -2.015886\n",
      "alr-1               -2.462368    18  0.537715  -1.327889  -3.596847\n",
      "ced-1               -3.257805    30  0.300106  -2.644019  -3.871590\n",
      "ced-12              -1.999798    12  0.508603  -0.880370  -3.119227\n",
      "cnd-1               -1.236586    18  0.312997  -0.576220  -1.896953\n",
      "dyf-11              -4.916347    30  0.240695  -4.424070  -5.408624\n",
      "gcy-18               0.099586     4  0.549377   1.847948  -1.648776\n",
      "gcy-23              -1.798092     9  0.446080  -0.769428  -2.826755\n",
      "gcy-8               -0.708515     9  0.304040  -0.007397  -1.409632\n",
      "lgc-34               0.035258     5  0.301361   0.871970  -0.801455\n",
      "mls-2               -3.354446    12  0.511554  -2.228524  -4.480368\n",
      "ngn-1                6.270735     3  4.191179  24.303924 -11.762455\n",
      "pde-1               -1.435166     9  0.435151  -0.431705  -2.438627\n",
      "psr-1               -1.622275     6  0.309012  -0.827934  -2.416616\n",
      "scrm-1              -1.174744     6  0.466424   0.024238  -2.373726\n",
      "tax-2               -1.237420    24  0.524650  -0.152099  -2.322741\n",
      "ttx-1               -1.892062     6  0.308589  -1.098809  -2.685315\n",
      "vab-3               -0.801894     5  0.654102   1.014183  -2.617972,                          Crab                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.000404  0.000809 -0.000809\n",
      "acd-1xdelm-2         0.003614     5  0.000990  0.006362  0.000865\n",
      "acd-1xdelm-2xdelm-1  0.003536     5  0.001015  0.006355  0.000717\n",
      "alr-1               -0.000251    18  0.000449  0.000696 -0.001198\n",
      "ced-1               -0.000683    30  0.000387  0.000109 -0.001475\n",
      "ced-12              -0.001626    12  0.000528 -0.000463 -0.002789\n",
      "cnd-1                0.000364    18  0.001728  0.004010 -0.003282\n",
      "dyf-11              -0.002645    30  0.000293 -0.002047 -0.003244\n",
      "gcy-18              -0.000095     4  0.000984  0.003035 -0.003226\n",
      "gcy-23              -0.002483     9  0.000504 -0.001321 -0.003646\n",
      "gcy-8               -0.001578     9  0.000589 -0.000219 -0.002937\n",
      "lgc-34               0.001591     5  0.001237  0.005026 -0.001845\n",
      "mls-2               -0.002527    12  0.000519 -0.001385 -0.003669\n",
      "ngn-1               -0.002448     3  0.001181  0.002632 -0.007527\n",
      "pde-1               -0.000106     9  0.000623  0.001331 -0.001543\n",
      "psr-1               -0.002106     6  0.000323 -0.001275 -0.002938\n",
      "scrm-1              -0.001839     6  0.001002  0.000737 -0.004415\n",
      "tax-2                0.001109    24  0.000543  0.002232 -0.000014\n",
      "ttx-1               -0.001907     6  0.000561 -0.000466 -0.003349\n",
      "vab-3                0.001459     5  0.001549  0.005759 -0.002841]\n",
      "[                    init_dura                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.040552  0.081236 -0.081236\n",
      "acd-1xdelm-2         0.480246     5  0.173986  0.963309 -0.002818\n",
      "acd-1xdelm-2xdelm-1  0.418246     5  0.150619  0.836430  0.000061\n",
      "alr-1               -0.918810    18  0.142141 -0.618918 -1.218702\n",
      "ced-1                0.110246    30  0.073665  0.260907 -0.040416\n",
      "ced-12               0.256746    12  0.178106  0.648754 -0.135263\n",
      "cnd-1               -1.509504    16  0.180382 -1.125028 -1.893980\n",
      "dyf-11               0.303801    30  0.150863  0.612351 -0.004749\n",
      "gcy-18               0.138412     4  0.062155  0.336216 -0.059391\n",
      "gcy-23               0.011746     9  0.089272  0.217607 -0.194115\n",
      "gcy-8                0.081468     9  0.099160  0.310131 -0.147196\n",
      "lgc-34              -0.230421     5  0.043990 -0.108286 -0.352557\n",
      "mls-2                0.377162    12  0.127342  0.657441  0.096884\n",
      "ngn-1               -2.495754     1       NaN       NaN       NaN\n",
      "pde-1               -0.157051     9  0.123261  0.127190 -0.441292\n",
      "psr-1                0.290079     6  0.034044  0.377593  0.202565\n",
      "scrm-1               0.601746     6  0.204926  1.128524  0.074967\n",
      "tax-2                0.600218    24  0.115883  0.839939  0.360496\n",
      "ttx-1                0.175912     6  0.066866  0.347798  0.004027\n",
      "vab-3                0.003246     5  0.059695  0.168986 -0.162494,                     init_prob                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.011641  0.023320 -0.023320\n",
      "acd-1xdelm-2         0.013676     5  0.024207  0.080885 -0.053533\n",
      "acd-1xdelm-2xdelm-1  0.023314     5  0.013417  0.060566 -0.013938\n",
      "alr-1               -0.248516    18  0.045049 -0.153470 -0.343561\n",
      "ced-1               -0.013331    30  0.022558  0.032805 -0.059467\n",
      "ced-12              -0.157524    12  0.058960 -0.027753 -0.287294\n",
      "cnd-1               -0.405751    16  0.042114 -0.315987 -0.495516\n",
      "dyf-11              -0.106397    30  0.041950 -0.020599 -0.192196\n",
      "gcy-18              -0.002536     4  0.018844  0.057433 -0.062505\n",
      "gcy-23              -0.100152     9  0.034435 -0.020744 -0.179560\n",
      "gcy-8               -0.124283     9  0.020314 -0.077439 -0.171128\n",
      "lgc-34              -0.036778     5  0.030586  0.048143 -0.121698\n",
      "mls-2               -0.055962    12  0.040983  0.034241 -0.146165\n",
      "ngn-1               -0.479981     1       NaN       NaN       NaN\n",
      "pde-1                0.002726     9  0.033748  0.080548 -0.075096\n",
      "psr-1               -0.023719     6  0.040122  0.079417 -0.126855\n",
      "scrm-1              -0.048140     6  0.054460  0.091854 -0.188133\n",
      "tax-2               -0.075388    24  0.027538 -0.018422 -0.132354\n",
      "ttx-1               -0.117404     6  0.031930 -0.035324 -0.199484\n",
      "vab-3                0.033236     5  0.032951  0.124721 -0.058250,                     init_speed                                    \n",
      "                          mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                              \n",
      "N2                    0.000000    57  0.004046  0.008105 -0.008105\n",
      "acd-1xdelm-2          0.032534     5  0.010057  0.060457  0.004611\n",
      "acd-1xdelm-2xdelm-1   0.032326     5  0.011372  0.063901  0.000751\n",
      "alr-1                -0.039915    18  0.007709 -0.023650 -0.056180\n",
      "ced-1                -0.025642    30  0.005022 -0.015372 -0.035913\n",
      "ced-12               -0.042852    12  0.010457 -0.019837 -0.065867\n",
      "cnd-1                -0.117187    16  0.012823 -0.089856 -0.144517\n",
      "dyf-11               -0.038555    30  0.007068 -0.024099 -0.053011\n",
      "gcy-18               -0.002256     4  0.006038  0.016958 -0.021470\n",
      "gcy-23               -0.031004     9  0.007356 -0.014041 -0.047967\n",
      "gcy-8                -0.028689     9  0.009497 -0.006790 -0.050589\n",
      "lgc-34               -0.003918     5  0.010745  0.025914 -0.033750\n",
      "mls-2                -0.053510    12  0.005867 -0.040596 -0.066424\n",
      "ngn-1                -0.025833     1       NaN       NaN       NaN\n",
      "pde-1                -0.014314     9  0.010016  0.008784 -0.037412\n",
      "psr-1                -0.003075     6  0.010450  0.023788 -0.029938\n",
      "scrm-1                0.007943     6  0.021753  0.063861 -0.047974\n",
      "tax-2                -0.034947    24  0.007035 -0.020394 -0.049500\n",
      "ttx-1                -0.034986     6  0.007489 -0.015734 -0.054238\n",
      "vab-3                 0.003516     5  0.010112  0.031593 -0.024560,                     final_dura                                    \n",
      "                          mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                              \n",
      "N2                    0.000000    57  0.034486  0.069083 -0.069083\n",
      "acd-1xdelm-2         -0.041845     5  0.135115  0.333294 -0.416984\n",
      "acd-1xdelm-2xdelm-1  -0.032401     5  0.059681  0.133300 -0.198101\n",
      "alr-1                 0.177414    18  0.073423  0.332323  0.022505\n",
      "ced-1                 0.005433    30  0.043589  0.094583 -0.083718\n",
      "ced-12               -0.012539    12  0.065491  0.131606 -0.156685\n",
      "cnd-1                -0.475039    16  0.099851 -0.262213 -0.687866\n",
      "dyf-11                0.093729    30  0.042835  0.181337  0.006121\n",
      "gcy-18               -0.493789     4  0.105263 -0.158795 -0.828784\n",
      "gcy-23                0.045099     9  0.082251  0.234771 -0.144572\n",
      "gcy-8                 0.044266     9  0.082816  0.235241 -0.146709\n",
      "lgc-34               -0.121289     5  0.287677  0.677429 -0.920008\n",
      "mls-2                -0.070595    12  0.079235  0.103800 -0.244990\n",
      "ngn-1                -0.979623     1       NaN       NaN       NaN\n",
      "pde-1                -0.030734     9  0.076380  0.145398 -0.206865\n",
      "psr-1                -0.011012     6  0.097548  0.239744 -0.261767\n",
      "scrm-1               -0.022678     6  0.034781  0.066729 -0.112085\n",
      "tax-2                -0.183651    24  0.058630 -0.062366 -0.304935\n",
      "ttx-1                 0.445655     6  0.127426  0.773213  0.118097\n",
      "vab-3                -0.293289     5  0.112855  0.020046 -0.606625,                     final_prob                                    \n",
      "                          mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                              \n",
      "N2                    0.000000    57  0.010202  0.020436 -0.020436\n",
      "acd-1xdelm-2         -0.048240     5  0.032296  0.041428 -0.137909\n",
      "acd-1xdelm-2xdelm-1  -0.012557     5  0.015681  0.030982 -0.056095\n",
      "alr-1                 0.014158    18  0.034578  0.087112 -0.058797\n",
      "ced-1                -0.003286    30  0.014672  0.026722 -0.033294\n",
      "ced-12                0.072355    12  0.028687  0.135495  0.009214\n",
      "cnd-1                -0.125975    16  0.011597 -0.101256 -0.150694\n",
      "dyf-11                0.087711    30  0.013655  0.115639  0.059783\n",
      "gcy-18               -0.106786     4  0.038080  0.014403 -0.227975\n",
      "gcy-23                0.077661     9  0.030667  0.148380  0.006942\n",
      "gcy-8                -0.016847     9  0.016450  0.021088 -0.054782\n",
      "lgc-34               -0.212915     5  0.012118 -0.179270 -0.246561\n",
      "mls-2                 0.064282    12  0.023757  0.116570  0.011993\n",
      "ngn-1                -0.237979     1       NaN       NaN       NaN\n",
      "pde-1                -0.046173     9  0.030487  0.024131 -0.116476\n",
      "psr-1                 0.093737     6  0.057165  0.240686 -0.053211\n",
      "scrm-1                0.245714     6  0.016926  0.289225  0.202204\n",
      "tax-2                 0.098270    24  0.026152  0.152370  0.044170\n",
      "ttx-1                -0.128303     6  0.018473 -0.080817 -0.175789\n",
      "vab-3                -0.012337     5  0.017374  0.035900 -0.060574,                     final_speed                                    \n",
      "                           mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                               \n",
      "N2                     0.000000    57  0.004412  0.008839 -0.008839\n",
      "acd-1xdelm-2           0.002539     5  0.013694  0.040561 -0.035483\n",
      "acd-1xdelm-2xdelm-1   -0.001421     5  0.008416  0.021945 -0.024787\n",
      "alr-1                 -0.009650    18  0.010225  0.011924 -0.031223\n",
      "ced-1                 -0.029294    30  0.005048 -0.018969 -0.039618\n",
      "ced-12                -0.013856    12  0.005749 -0.001202 -0.026511\n",
      "cnd-1                 -0.051227    16  0.019747 -0.009137 -0.093318\n",
      "dyf-11                -0.015145    30  0.004435 -0.006074 -0.024216\n",
      "gcy-18                -0.049696     4  0.005104 -0.033454 -0.065938\n",
      "gcy-23                -0.004058     9  0.005817  0.009357 -0.017472\n",
      "gcy-8                 -0.006772     9  0.006425  0.008043 -0.021588\n",
      "lgc-34                -0.017603     5  0.023696  0.048187 -0.083394\n",
      "mls-2                 -0.042325    12  0.007635 -0.025522 -0.059129\n",
      "ngn-1                 -0.128557     1       NaN       NaN       NaN\n",
      "pde-1                 -0.010185     9  0.012259  0.018084 -0.038454\n",
      "psr-1                 -0.006308     6  0.007407  0.012733 -0.025350\n",
      "scrm-1                 0.016447     6  0.016479  0.058808 -0.025913\n",
      "tax-2                 -0.040903    24  0.003761 -0.033123 -0.048682\n",
      "ttx-1                 -0.016284     6  0.006765  0.001106 -0.033675\n",
      "vab-3                 -0.007345     5  0.003334  0.001911 -0.016601,                     habit_dura                                    \n",
      "                          mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                              \n",
      "N2                    0.000000    57  0.049598  0.099356 -0.099356\n",
      "acd-1xdelm-2          0.522091     5  0.044263  0.644983  0.399198\n",
      "acd-1xdelm-2xdelm-1   0.450646     5  0.151454  0.871150  0.030142\n",
      "alr-1                -1.096224    18  0.139324 -0.802277 -1.390171\n",
      "ced-1                 0.104813    30  0.084845  0.278341 -0.068716\n",
      "ced-12                0.269285    12  0.174662  0.653713 -0.115142\n",
      "cnd-1                -1.034465    16  0.140887 -0.734171 -1.334759\n",
      "dyf-11                0.210072    30  0.157637  0.532475 -0.112331\n",
      "gcy-18                0.632202     4  0.121429  1.018642  0.245762\n",
      "gcy-23               -0.033354     9  0.086867  0.166962 -0.233670\n",
      "gcy-8                 0.037202     9  0.139163  0.358112 -0.283709\n",
      "lgc-34               -0.109132     5  0.287597  0.689366 -0.907629\n",
      "mls-2                 0.447757    12  0.174650  0.832160  0.063355\n",
      "ngn-1                -1.516132     1       NaN       NaN       NaN\n",
      "pde-1                -0.126317     9  0.144659  0.207268 -0.459901\n",
      "psr-1                 0.301091     6  0.106767  0.575544  0.026637\n",
      "scrm-1                0.624424     6  0.228186  1.210996  0.037852\n",
      "tax-2                 0.783868    24  0.149802  1.093757  0.473980\n",
      "ttx-1                -0.269743     6  0.179721  0.192245 -0.731731\n",
      "vab-3                 0.296535     5  0.162442  0.747546 -0.154476,                     habit_prob                                    \n",
      "                          mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                              \n",
      "N2                    0.000000    57  0.013594  0.027231 -0.027231\n",
      "acd-1xdelm-2          0.061916     5  0.052221  0.206905 -0.083073\n",
      "acd-1xdelm-2xdelm-1   0.035871     5  0.016484  0.081638 -0.009897\n",
      "alr-1                -0.262673    18  0.036874 -0.184875 -0.340471\n",
      "ced-1                -0.010045    30  0.026989  0.045153 -0.065243\n",
      "ced-12               -0.229878    12  0.065368 -0.086004 -0.373753\n",
      "cnd-1                -0.279776    16  0.038126 -0.198512 -0.361041\n",
      "dyf-11               -0.194109    30  0.045350 -0.101358 -0.286859\n",
      "gcy-18                0.104251     4  0.036989  0.221967 -0.013466\n",
      "gcy-23               -0.177813     9  0.047747 -0.067707 -0.287919\n",
      "gcy-8                -0.107436     9  0.029822 -0.038667 -0.176205\n",
      "lgc-34                0.176138     5  0.031849  0.264564  0.087711\n",
      "mls-2                -0.120244    12  0.041551 -0.028791 -0.211696\n",
      "ngn-1                -0.242001     1       NaN       NaN       NaN\n",
      "pde-1                 0.048899     9  0.039885  0.140875 -0.043077\n",
      "psr-1                -0.117457     6  0.092208  0.119572 -0.354485\n",
      "scrm-1               -0.293854     6  0.069638 -0.114845 -0.472863\n",
      "tax-2                -0.173658    24  0.033077 -0.105234 -0.242082\n",
      "ttx-1                 0.010899     6  0.039117  0.111453 -0.089655\n",
      "vab-3                 0.045573     5  0.043856  0.167335 -0.076190,                     habit_speed                                    \n",
      "                           mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                               \n",
      "N2                     0.000000    57  0.004309  0.008632 -0.008632\n",
      "acd-1xdelm-2           0.029995     5  0.015646  0.073435 -0.013445\n",
      "acd-1xdelm-2xdelm-1    0.033747     5  0.007510  0.054597  0.012896\n",
      "alr-1                 -0.030266    18  0.009108 -0.011049 -0.049482\n",
      "ced-1                  0.003651    30  0.004128  0.012094 -0.004792\n",
      "ced-12                -0.028995    12  0.009057 -0.009060 -0.048930\n",
      "cnd-1                 -0.065959    16  0.012815 -0.038644 -0.093274\n",
      "dyf-11                -0.023410    30  0.005668 -0.011819 -0.035001\n",
      "gcy-18                 0.047440     4  0.009116  0.076450  0.018429\n",
      "gcy-23                -0.026947     9  0.005378 -0.014544 -0.039349\n",
      "gcy-8                 -0.021917     9  0.007176 -0.005369 -0.038465\n",
      "lgc-34                 0.013685     5  0.022043  0.074887 -0.047517\n",
      "mls-2                 -0.011184    12  0.004621 -0.001014 -0.021355\n",
      "ngn-1                  0.102724     1       NaN       NaN       NaN\n",
      "pde-1                 -0.004129     9  0.009154  0.016979 -0.025238\n",
      "psr-1                  0.003233     6  0.013058  0.036800 -0.030334\n",
      "scrm-1                -0.008504     6  0.012054  0.022481 -0.039490\n",
      "tax-2                  0.005956    24  0.005072  0.016448 -0.004536\n",
      "ttx-1                 -0.018701     6  0.009036  0.004526 -0.041929\n",
      "vab-3                  0.010861     5  0.007285  0.031089 -0.009366,                          time                                    \n",
      "                         mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                             \n",
      "N2                   0.000000    57  0.001952  0.003910 -0.003910\n",
      "acd-1xdelm-2         0.004029     5  0.005722  0.019915 -0.011857\n",
      "acd-1xdelm-2xdelm-1  0.007362     5  0.004768  0.020601 -0.005877\n",
      "alr-1                0.007490    18  0.002712  0.013211  0.001768\n",
      "ced-1                0.001879    30  0.003007  0.008029 -0.004271\n",
      "ced-12               0.001754    12  0.003467  0.009385 -0.005878\n",
      "cnd-1                0.006629    16  0.004216  0.015615 -0.002358\n",
      "dyf-11               0.003490    30  0.002497  0.008596 -0.001617\n",
      "gcy-18               0.003920     4  0.008951  0.032408 -0.024567\n",
      "gcy-23               0.001156     9  0.005832  0.014606 -0.012293\n",
      "gcy-8                0.006073     9  0.004196  0.015749 -0.003603\n",
      "lgc-34               0.015295     5  0.003658  0.025450  0.005140\n",
      "mls-2                0.002420    12  0.004445  0.012204 -0.007363\n",
      "ngn-1                0.024629     1       NaN       NaN       NaN\n",
      "pde-1               -0.000649     9  0.004827  0.010481 -0.011780\n",
      "psr-1               -0.008288     6  0.005755  0.006506 -0.023083\n",
      "scrm-1               0.010712     6  0.005807  0.025640 -0.004216\n",
      "tax-2                0.004476    24  0.002297  0.009227 -0.000276\n",
      "ttx-1                0.008379     6  0.005746  0.023148 -0.006391\n",
      "vab-3                0.011629     5  0.002761  0.019295  0.003962,                     taps                           \n",
      "                    mean count  sem ci95_hi ci95_lo\n",
      "Gene                                               \n",
      "N2                   0.0    57  0.0     NaN     NaN\n",
      "acd-1xdelm-2         0.0     5  0.0     NaN     NaN\n",
      "acd-1xdelm-2xdelm-1  0.0     5  0.0     NaN     NaN\n",
      "alr-1                0.0    18  0.0     NaN     NaN\n",
      "ced-1                0.0    30  0.0     NaN     NaN\n",
      "ced-12               0.0    12  0.0     NaN     NaN\n",
      "cnd-1                0.0    16  0.0     NaN     NaN\n",
      "dyf-11               0.0    30  0.0     NaN     NaN\n",
      "gcy-18               0.0     4  0.0     NaN     NaN\n",
      "gcy-23               0.0     9  0.0     NaN     NaN\n",
      "gcy-8                0.0     9  0.0     NaN     NaN\n",
      "lgc-34               0.0     5  0.0     NaN     NaN\n",
      "mls-2                0.0    12  0.0     NaN     NaN\n",
      "ngn-1                0.0     1  NaN     NaN     NaN\n",
      "pde-1                0.0     9  0.0     NaN     NaN\n",
      "psr-1                0.0     6  0.0     NaN     NaN\n",
      "scrm-1               0.0     6  0.0     NaN     NaN\n",
      "tax-2                0.0    24  0.0     NaN     NaN\n",
      "ttx-1                0.0     6  0.0     NaN     NaN\n",
      "vab-3                0.0     5  0.0     NaN     NaN,                     recovery_dura                                    \n",
      "                             mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                 \n",
      "N2                       0.000000    57  0.060919  0.122035 -0.122035\n",
      "acd-1xdelm-2            -0.079041     5  0.144921  0.323324 -0.481406\n",
      "acd-1xdelm-2xdelm-1      0.002626     5  0.175446  0.489742 -0.484490\n",
      "alr-1                   -0.175819    18  0.107845  0.051714 -0.403351\n",
      "ced-1                    0.283459    30  0.089199  0.465892  0.101027\n",
      "ced-12                   0.109459    12  0.234229  0.624993 -0.406075\n",
      "cnd-1                   -0.175333    16  0.146669  0.137285 -0.487951\n",
      "dyf-11                  -0.054819    30  0.117374  0.185237 -0.294875\n",
      "gcy-18                   0.734876     4  0.145627  1.198325  0.271427\n",
      "gcy-23                  -0.024430     9  0.117815  0.247251 -0.296111\n",
      "gcy-8                   -0.019152     9  0.072497  0.148027 -0.186331\n",
      "lgc-34                   0.420292     5  0.224800  1.044437 -0.203852\n",
      "mls-2                   -0.105958    12  0.117763  0.153238 -0.365153\n",
      "ngn-1                   -3.274708     1       NaN       NaN       NaN\n",
      "pde-1                    0.125663     9  0.158617  0.491435 -0.240109\n",
      "psr-1                   -0.260541     6  0.125522  0.062123 -0.583205\n",
      "scrm-1                  -0.094708     6  0.176687  0.359481 -0.548896\n",
      "tax-2                    0.516404    24  0.125481  0.775980  0.256827\n",
      "ttx-1                    0.052792     6  0.149665  0.437518 -0.331933\n",
      "vab-3                    0.096292     5  0.120876  0.431898 -0.239313,                     recovery_prob                                    \n",
      "                             mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                 \n",
      "N2                       0.000000    57  0.017317  0.034690 -0.034690\n",
      "acd-1xdelm-2            -0.058777     5  0.040302  0.053120 -0.170673\n",
      "acd-1xdelm-2xdelm-1     -0.017213     5  0.045995  0.110489 -0.144914\n",
      "alr-1                   -0.052354    18  0.035009  0.021508 -0.126216\n",
      "ced-1                   -0.000709    30  0.031898  0.064530 -0.065949\n",
      "ced-12                  -0.139533    12  0.047644 -0.034669 -0.244398\n",
      "cnd-1                    0.010789    16  0.051278  0.120084 -0.098507\n",
      "dyf-11                  -0.163130    30  0.047910 -0.065144 -0.261116\n",
      "gcy-18                  -0.030235     4  0.046508  0.117772 -0.178243\n",
      "gcy-23                  -0.227903     9  0.046413 -0.120874 -0.334933\n",
      "gcy-8                   -0.185993     9  0.030782 -0.115010 -0.256976\n",
      "lgc-34                   0.323643     5  0.047790  0.456328  0.190957\n",
      "mls-2                   -0.090732    12  0.033832 -0.016268 -0.165196\n",
      "ngn-1                    0.154007     1       NaN       NaN       NaN\n",
      "pde-1                    0.073191     9  0.048491  0.185011 -0.038629\n",
      "psr-1                   -0.186740     6  0.050067 -0.058038 -0.315441\n",
      "scrm-1                  -0.184684     6  0.054295 -0.045115 -0.324253\n",
      "tax-2                    0.031631    24  0.032795  0.099474 -0.036212\n",
      "ttx-1                    0.077721     6  0.065247  0.245443 -0.090001\n",
      "vab-3                   -0.086295     5  0.052145  0.058484 -0.231074,                     recovery_speed                                    \n",
      "                              mean count       sem   ci95_hi   ci95_lo\n",
      "Gene                                                                  \n",
      "N2                        0.000000    57  0.003335  0.006681 -0.006681\n",
      "acd-1xdelm-2             -0.003389     5  0.009983  0.024327 -0.031105\n",
      "acd-1xdelm-2xdelm-1       0.005231     5  0.008944  0.030064 -0.019602\n",
      "alr-1                     0.026642    18  0.006374  0.040090  0.013194\n",
      "ced-1                     0.012802    30  0.004871  0.022764  0.002840\n",
      "ced-12                   -0.009493    12  0.005328  0.002233 -0.021220\n",
      "cnd-1                     0.015569    16  0.009622  0.036077 -0.004940\n",
      "dyf-11                   -0.002039    30  0.004988  0.008162 -0.012240\n",
      "gcy-18                   -0.004572     4  0.014768  0.042428 -0.051571\n",
      "gcy-23                   -0.038874     9  0.006544 -0.023783 -0.053965\n",
      "gcy-8                    -0.038180     9  0.002767 -0.031800 -0.044560\n",
      "lgc-34                    0.003027     5  0.010014  0.030829 -0.024776\n",
      "mls-2                     0.003524    12  0.007199  0.019370 -0.012321\n",
      "ngn-1                     0.087972     1       NaN       NaN       NaN\n",
      "pde-1                    -0.000155     9  0.006362  0.014516 -0.014826\n",
      "psr-1                    -0.028322     6  0.011554  0.001380 -0.058024\n",
      "scrm-1                   -0.010154     6  0.007972  0.010338 -0.030645\n",
      "tax-2                     0.021303    24  0.004721  0.031069  0.011537\n",
      "ttx-1                    -0.004890     6  0.003062  0.002981 -0.012762\n",
      "vab-3                    -0.006369     5  0.011027  0.024248 -0.036985]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:2351: RuntimeWarning: invalid value encountered in multiply\n",
      "  lower_bound = _a * scale + loc\n",
      "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:2352: RuntimeWarning: invalid value encountered in multiply\n",
      "  upper_bound = _b * scale + loc\n"
     ]
    }
   ],
   "source": [
    "# test1 = extract_phenotypes(baseline_output_byplate)\n",
    "# print(test1)\n",
    "list_baseline_MSD=calculate_MSD(extract_phenotypes(baseline_output_byplate))\n",
    "print(list_baseline_MSD)\n",
    "# test2 = extract_phenotypes(tap_data_byplate)\n",
    "# print(test2)\n",
    "list_tap_MSD=calculate_MSD(extract_phenotypes(tap_data_byplate))\n",
    "print(list_tap_MSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e77ef7a-0c58-424a-ab6b-3008aa9c92c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in list_baseline_MSD:\n",
    "    # b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if a.columns.levels[0] == list_baseline_MSD[0].columns.levels[0]:\n",
    "        baseline_MSD=a\n",
    "    else:\n",
    "        baseline_MSD=baseline_MSD.join(a)\n",
    "\n",
    "for a in list_tap_MSD:\n",
    "    # b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if a.columns.levels[0] == list_tap_MSD[0].columns.levels[0]:\n",
    "        tap_MSD=a\n",
    "    else:\n",
    "        tap_MSD=tap_MSD.join(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95142e9d-8f20-4842-8f72-8bcfa7711836",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(baseline_MSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b728b322-701d-4af7-98e6-499ce59edcf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tap_MSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff9b3b8-f2cd-4562-b17f-f8c6c446086f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(baseline_MSD.columns.levels[0])\n",
    "print(\"-\"*30)\n",
    "print(tap_MSD.columns.levels[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7256140-6cee-4717-8b86-5263e6748971",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df=list_baseline_MSD[-1]\n",
    "# test_df['Crab']\n",
    "test_df.columns.levels[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f26e6087-8576-4e49-b350-59491a231718",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_MSD = baseline_MSD.join(tap_MSD, on='Gene')\n",
    "print(combined_MSD.columns.levels[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bcdfea7-eb0a-4e8a-9a24-33212ce25c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_MSD=combined_MSD.rename(columns={\"habit_dura\":\"Habituation of Response Duration\",\n",
    "                                         \"habit_prob\": \"Habituation of Respones Probability\",\n",
    "                                         \"habit_speed\":\"Habituation of Response Speed\",\n",
    "                                         \"init_dura\": \"Initial Response Duration\",\n",
    "                                         \"init_prob\": \"Initial Response Probability\",\n",
    "                                         \"init_speed\": \"Initial Response Speed\",\n",
    "                                         \"final_dura\": \"Final Response Duration\",\n",
    "                                         \"final_prob\": \"Final Response Probability\",\n",
    "                                         \"final_speed\": \"Final Response Speed\",\n",
    "                                         \"recovery_dura\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                         \"recovery_prob\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                         \"recovery_speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "print(combined_MSD.columns.levels[0])\n",
    "print(len(combined_MSD.columns.levels[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c83136c9-ee5b-463b-ad39-0e5403199cd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_MSD=combined_MSD.reset_index()\n",
    "combined_MSD.columns = combined_MSD.columns.to_flat_index().str.join('-')\n",
    "combined_MSD=combined_MSD.rename(columns={\"Gene-\": \"Gene\"})\n",
    "combined_MSD['Screen']=Screen\n",
    "print(combined_MSD)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18e39002-cc71-4b7e-8404-6fd5d9eaccef",
   "metadata": {},
   "source": [
    "## Allele-level SMD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "167044c8-add0-4301-a0e8-e76f7ed289ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(baseline_output_allele_byplate[baseline_output_allele_byplate.dataset=='N2_XJ1'])\n",
    "# print(tap_data_allele_byplate[tap_data_allele_byplate.dataset=='N2_N2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dff3dfa0-942c-4213-801f-201d0f1ce035",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test1=extract_allele_phenotypes(baseline_output_allele_byplate)\n",
    "# print(test1)\n",
    "list_allele_baseline_MSD=calculate_allele_MSD(extract_allele_phenotypes(baseline_output_allele_byplate))\n",
    "print(list_allele_baseline_MSD)\n",
    "\n",
    "# test2=extract_allele_phenotypes(tap_data_allele_byplate)\n",
    "# print(test2[0][test2[0].dataset=='N2_N2'])\n",
    "list_allele_tap_MSD=calculate_allele_MSD(extract_allele_phenotypes(tap_data_allele_byplate))\n",
    "print(list_allele_tap_MSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc3efb1d-2dc5-4451-8969-48ac780c9395",
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in list_allele_baseline_MSD:\n",
    "    # b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if a.columns.levels[0] == list_allele_baseline_MSD[0].columns.levels[0]:\n",
    "        allele_baseline_MSD=a\n",
    "    else:\n",
    "        allele_baseline_MSD=allele_baseline_MSD.join(a)\n",
    "\n",
    "for a in list_allele_tap_MSD:\n",
    "    # b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if a.columns.levels[0] == list_allele_tap_MSD[0].columns.levels[0]:\n",
    "        allele_tap_MSD=a\n",
    "    else:\n",
    "        allele_tap_MSD=allele_tap_MSD.join(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e80ccd78-c5db-4e4d-987a-b7cd148d795e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(allele_baseline_MSD.columns.levels[0])\n",
    "print(\"-\"*30)\n",
    "print(allele_tap_MSD.columns.levels[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b733fd0a-041d-442c-b895-43a566a1b17a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eecea70c-634e-4069-8dc1-fcb32a9900c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_combined_MSD = allele_baseline_MSD.join(allele_tap_MSD, on='dataset')\n",
    "print(allele_combined_MSD.columns.levels[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6020d04d-07b5-4420-8a5f-884c67ef4146",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_combined_MSD=allele_combined_MSD.rename(columns={\"habit_dura\":\"Habituation of Response Duration\",\n",
    "                                                         \"habit_prob\": \"Habituation of Respones Probability\",\n",
    "                                                         \"habit_speed\":\"Habituation of Response Speed\",\n",
    "                                                         \"init_dura\": \"Initial Response Duration\",\n",
    "                                                         \"init_prob\": \"Initial Response Probability\",\n",
    "                                                         \"init_speed\": \"Initial Response Speed\",\n",
    "                                                         \"final_dura\": \"Final Response Duration\",\n",
    "                                                         \"final_prob\": \"Final Response Probability\",\n",
    "                                                         \"final_speed\": \"Final Response Speed\",\n",
    "                                                         \"recovery_dura\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                                         \"recovery_prob\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                                         \"recovery_speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "print(allele_combined_MSD.columns.levels[0])\n",
    "print(len(allele_combined_MSD.columns.levels[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6657aae5-3713-4240-95a5-dfca119d6199",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_combined_MSD=allele_combined_MSD.reset_index()\n",
    "allele_combined_MSD.columns = allele_combined_MSD.columns.to_flat_index().str.join('-')\n",
    "allele_combined_MSD=allele_combined_MSD.rename(columns={\"dataset-\": \"dataset\"})\n",
    "allele_combined_MSD['Screen']=Screen\n",
    "print(allele_combined_MSD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff34bdf6-0f9a-42d5-9df1-51f28a6ee777",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d136092-c389-4762-bfb4-c2d378a43bdf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449c8711-858c-48af-b259-66546f8755cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f575a47-8b29-404a-9147-203d21aec5a9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b5784896-b993-4ecb-9ae9-c168d7469d90",
   "metadata": {},
   "source": [
    "### Allele-level T-stat analysis of baseline data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "827bd865-1ee0-4c89-b7ec-bf079f7688ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "list=baseline_output[\"dataset\"].unique()\n",
    "print(list)\n",
    "print(len(list))\n",
    "print(baseline_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e62abf35-b324-4b96-9328-91df56e05470",
   "metadata": {},
   "outputs": [],
   "source": [
    "def TTest_Allele(Type, DF_ref, output):\n",
    "    for a in DF_ref['dataset'].unique():\n",
    "        # Tstat=ttest_ind(DF_ref[DF_ref.dataset == a][Type], DF_ref[DF_ref.Allele == \"XJ1\" or DF_ref.Allele == \"N2\"][Type],equal_var=False)[0]\n",
    "        Tstat=ttest_ind(DF_ref[DF_ref.dataset == a][Type], DF_ref[DF_ref.Allele.isin([\"XJ1\",\"N2\"])][Type],equal_var=False)[0]\n",
    "        row=[a, Tstat]\n",
    "        output.loc[len(output)]=row\n",
    "    print(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc477df9-91aa-41d6-a527-c61d20f99fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_instantspeed_T_allele=pd.DataFrame(columns = [\"dataset\",\"Instantaneous Speed\"])\n",
    "PD_baseline_intspeed_T_allele=pd.DataFrame(columns = [\"dataset\",\"Interval Speed\"])\n",
    "PD_baseline_bias_T_allele=pd.DataFrame(columns = [\"dataset\",\"Bias\"])\n",
    "PD_baseline_width_T_allele=pd.DataFrame(columns = [\"dataset\",\"Width\"])\n",
    "PD_baseline_length_T_allele=pd.DataFrame(columns = [\"dataset\",\"Length\"])\n",
    "PD_baseline_area_T_allele=pd.DataFrame(columns = [\"dataset\",\"Area\"])\n",
    "PD_baseline_angularspeed_T_allele=pd.DataFrame(columns = [\"dataset\",\"Angular Speed\"])\n",
    "PD_baseline_aspectratio_T_allele=pd.DataFrame(columns = [\"dataset\",\"Aspect Ratio\"])\n",
    "PD_baseline_kink_T_allele=pd.DataFrame(columns = [\"dataset\",\"Kink\"])\n",
    "PD_baseline_curve_T_allele=pd.DataFrame(columns = [\"dataset\",\"Curve\"])\n",
    "PD_baseline_crab_T_allele=pd.DataFrame(columns = [\"dataset\",\"Crab\"])\n",
    "\n",
    "list_baseline_Tstats_allele=[PD_baseline_instantspeed_T_allele,\n",
    "                      PD_baseline_intspeed_T_allele,\n",
    "                      PD_baseline_bias_T_allele,\n",
    "                      PD_baseline_width_T_allele,\n",
    "                      PD_baseline_length_T_allele,\n",
    "                      PD_baseline_area_T_allele,\n",
    "                      PD_baseline_angularspeed_T_allele,\n",
    "                      PD_baseline_aspectratio_T_allele,\n",
    "                      PD_baseline_kink_T_allele,\n",
    "                      PD_baseline_curve_T_allele,\n",
    "                      PD_baseline_crab_T_allele]\n",
    "\n",
    "list_baseline_metrics=[\"Instantaneous Speed\",\n",
    "                       \"Interval Speed\",\n",
    "                       \"Bias\",\n",
    "                       \"Width\",\n",
    "                       \"Length\",\n",
    "                       \"Area\",\n",
    "                       \"Angular Speed\",\n",
    "                       \"Aspect Ratio\",\n",
    "                       \"Kink\",\n",
    "                       \"Curve\",\n",
    "                       \"Crab\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b91e9cb8-ac8c-4fb1-a687-4e6017d3c296",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tqdm\n",
    "for x in baseline_output['dataset'].unique():\n",
    "    # if x == \"N2_XJ1\" or x == \"N2_N2\":\n",
    "    if x == \"N2\":\n",
    "        pass\n",
    "    else:\n",
    "        baseline_output_gene=baseline_output[baseline_output['dataset']==x]\n",
    "        gene_baseline_data=baseline_output[baseline_output['Date'].isin(baseline_output_gene['Date'].unique())]\n",
    "        gene_baseline_data_final=gene_baseline_data[gene_baseline_data['dataset'].isin(['N2',x])]\n",
    "        # gene_baseline_data_final=gene_baseline_data[gene_baseline_data['dataset'].isin(['N2_XJ1', 'N2_N2',x])]\n",
    "        # print(gene_baseline_data_final['Gene'].unique())\n",
    "        for a,b in zip(list_baseline_metrics, list_baseline_Tstats_allele):\n",
    "            TTest_Allele(a, gene_baseline_data_final, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e2655b3-5fe2-4d9d-b16c-ea50f807911c",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_Tstats_allele=pd.DataFrame()\n",
    "for a in list_baseline_Tstats_allele:\n",
    "    b=a.groupby(['dataset'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if b.columns.values[1] == list_baseline_Tstats_allele[0].columns.values[1]:\n",
    "        PD_baseline_Tstats_allele=b\n",
    "    else:\n",
    "        PD_baseline_Tstats_allele=PD_baseline_Tstats_allele.join(b.iloc[:,1])\n",
    "        \n",
    "print(PD_baseline_Tstats_allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bcd6102-56f2-4638-a988-207d59fbf4b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_Tstats_allele=PD_baseline_Tstats_allele.set_index(\"dataset\")\n",
    "print(PD_baseline_Tstats_allele)\n",
    "PD_baseline_Tstats_allele_sort=PD_baseline_Tstats_allele.sort_index()\n",
    "print(PD_baseline_Tstats_allele_sort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8b6e5cc-a2b0-4fd8-9bc4-668b1cc22d5e",
   "metadata": {},
   "source": [
    "### Gene-level T-stat analysis of baseline data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f2c5731-43cb-4b21-991c-15e85684c817",
   "metadata": {},
   "outputs": [],
   "source": [
    "def TTest_Gene(Type, DF_ref, output):\n",
    "    for a in DF_ref['Gene'].unique():\n",
    "        Tstat=ttest_ind(DF_ref[DF_ref.Gene == a][Type], DF_ref[DF_ref.Gene == \"N2\"][Type],equal_var=False)[0]\n",
    "        row=[a, Tstat]\n",
    "        output.loc[len(output)]=row\n",
    "    print(output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803ea9e3-1121-4b1c-ac15-77bcef16cd1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_instantspeed_T=pd.DataFrame(columns = [\"Gene\",\"Instantaneous Speed\"])\n",
    "PD_baseline_intspeed_T=pd.DataFrame(columns = [\"Gene\",\"Interval Speed\"])\n",
    "PD_baseline_bias_T=pd.DataFrame(columns = [\"Gene\",\"Bias\"])\n",
    "PD_baseline_width_T=pd.DataFrame(columns = [\"Gene\",\"Width\"])\n",
    "PD_baseline_length_T=pd.DataFrame(columns = [\"Gene\",\"Length\"])\n",
    "PD_baseline_area_T=pd.DataFrame(columns = [\"Gene\",\"Area\"])\n",
    "PD_baseline_angularspeed_T=pd.DataFrame(columns = [\"Gene\",\"Angular Speed\"])\n",
    "PD_baseline_aspectratio_T=pd.DataFrame(columns = [\"Gene\",\"Aspect Ratio\"])\n",
    "PD_baseline_kink_T=pd.DataFrame(columns = [\"Gene\",\"Kink\"])\n",
    "PD_baseline_curve_T=pd.DataFrame(columns = [\"Gene\",\"Curve\"])\n",
    "PD_baseline_crab_T=pd.DataFrame(columns = [\"Gene\",\"Crab\"])\n",
    "\n",
    "list_baseline_Tstats=[PD_baseline_instantspeed_T,\n",
    "                      PD_baseline_intspeed_T,\n",
    "                      PD_baseline_bias_T,\n",
    "                      PD_baseline_width_T,\n",
    "                      PD_baseline_length_T,\n",
    "                      PD_baseline_area_T,\n",
    "                      PD_baseline_angularspeed_T,\n",
    "                      PD_baseline_aspectratio_T,\n",
    "                      PD_baseline_kink_T,\n",
    "                      PD_baseline_curve_T,\n",
    "                      PD_baseline_crab_T]\n",
    "\n",
    "list_baseline_metrics=[\"Instantaneous Speed\",\n",
    "                       \"Interval Speed\",\n",
    "                       \"Bias\",\n",
    "                       \"Width\",\n",
    "                       \"Length\",\n",
    "                       \"Area\",\n",
    "                       \"Angular Speed\",\n",
    "                       \"Aspect Ratio\",\n",
    "                       \"Kink\",\n",
    "                       \"Curve\",\n",
    "                       \"Crab\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfea2d84-fe76-4f97-b0fd-e17e3311cd73",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in baseline_output['Gene'].unique():\n",
    "    if x == \"N2\":\n",
    "        pass\n",
    "    else:\n",
    "        baseline_output_gene=baseline_output[baseline_output['Gene']==x]\n",
    "        gene_baseline_data=baseline_output[baseline_output['Date'].isin(baseline_output_gene['Date'].unique())]\n",
    "        gene_baseline_data_final=gene_baseline_data[gene_baseline_data['Gene'].isin(['N2', x])]\n",
    "        # print(gene_baseline_data_final['Gene'].unique())\n",
    "        for a,b in zip(list_baseline_metrics, list_baseline_Tstats):\n",
    "            TTest_Gene(a, gene_baseline_data_final, b)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01becf6-c0f6-444a-b9dc-378e14621b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_Tstats=pd.DataFrame()\n",
    "for a in list_baseline_Tstats:\n",
    "    b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if b.columns.values[1] == list_baseline_Tstats[0].columns.values[1]:\n",
    "        PD_baseline_Tstats=b\n",
    "    else:\n",
    "        PD_baseline_Tstats=PD_baseline_Tstats.join(b.iloc[:,1])\n",
    "        \n",
    "print(PD_baseline_Tstats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3fce379-93d5-4234-9348-6e51862a284e",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_baseline_Tstats=PD_baseline_Tstats.set_index(\"Gene\")\n",
    "print(PD_baseline_Tstats)\n",
    "PD_baseline_Tstats_sort=PD_baseline_Tstats.sort_index()\n",
    "print(PD_baseline_Tstats_sort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80323af7-e75f-4678-b5dd-810a7607fc83",
   "metadata": {},
   "source": [
    "### T-stat analysis for tap-response data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2906984d-b3a7-40d2-856b-5ca4672d629f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_tap=tap_output\n",
    "# PD_first_tap = PD_tap[(PD_tap.taps==1)].reset_index().drop(columns=\"index\")\n",
    "# PD_last_three_taps = PD_tap[((PD_tap.taps>=28)&(PD_tap.taps<=30))].reset_index().drop(columns=\"index\")\n",
    "# PD_recov_taps = PD_tap[(PD_tap.taps==31)].reset_index().drop(columns=\"index\")\n",
    "# print(PD_first_tap)\n",
    "# print(PD_last_three_taps)\n",
    "# print(PD_recov_taps)\n",
    "# PD_final_taps=PD_last_three_taps.groupby([\"dataset\", \"Date\",\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"plate\"]).mean().reset_index()\n",
    "# print(PD_final_taps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "862c64cc-d069-4e18-b337-f4820a958cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_final_taps=PD_final_taps.rename(columns={\"dura\": \"final_dura\", \"prob\": \"final_prob\", \"speed\": \"final_speed\"}, errors=\"raise\")\n",
    "# PD_first_tap=PD_first_tap.rename(columns={\"dura\": \"init_dura\", \"prob\": \"init_prob\", \"speed\": \"init_speed\"}, errors=\"raise\")\n",
    "# print(PD_final_taps)\n",
    "# print(PD_first_tap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82dc9c5f-3fb7-47af-a0da-6ba8eaa0b7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_habit_levels = pd.merge(PD_first_tap, PD_final_taps, on =['dataset','plate',\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"Date\"], how ='left')\n",
    "# PD_habit_levels = PD_habit_levels.dropna()\n",
    "# print(PD_habit_levels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cc1694f-aaf9-4ba0-95c6-3ce98480f298",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_habit_levels['habit_dura']=PD_habit_levels.init_dura-PD_habit_levels.final_dura\n",
    "# PD_habit_levels['habit_prob']=PD_habit_levels.init_prob-PD_habit_levels.final_prob\n",
    "# PD_habit_levels['habit_speed']=PD_habit_levels.init_speed-PD_habit_levels.final_speed\n",
    "# # PD_habit_levels=PD_habit_levels.drop(columns=['time_x','time_y','dist_x','dist_y', 'Allele','Gene'])\n",
    "# PD_habit_levels=PD_habit_levels.drop(columns=['time_x','time_y','dist_x','dist_y'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f236ade4-f79c-4f64-aba2-366ef1d41318",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_recov_taps=PD_recov_taps.rename(columns={\"dura\": \"recov_dura\", \"prob\": \"recov_prob\", \"speed\":\"recov_speed\"})\n",
    "# PD_habituation = pd.merge(PD_habit_levels, PD_recov_taps, on =['dataset','plate',\"Plate_id\",\"Screen\",\"Gene\",\"Allele\",\"Date\"], how ='left')\n",
    "# PD_habituation = PD_habituation.drop(columns=['taps_x','taps_y','taps','time'])\n",
    "# PD_habituation = PD_habituation.dropna()\n",
    "# print(PD_habituation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1e2a1e-8664-43e8-a8af-48f1ebbf3585",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PD_habituation['recovery_dura']=PD_habituation.init_dura-PD_habituation.recov_dura\n",
    "# PD_habituation['recovery_prob']=PD_habituation.init_prob-PD_habituation.recov_prob\n",
    "# PD_habituation['recovery_speed']=PD_habituation.init_speed-PD_habituation.recov_speed\n",
    "# PD_habituation=PD_habituation.dropna()\n",
    "# print(PD_habituation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d945bd9-912e-4dc9-926d-7d04585a8a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tap_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da90e6af-cfae-447c-9716-bcdeee522764",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tap_data.columns.values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae6c679a-166c-43d0-946e-226ee9a20c1d",
   "metadata": {},
   "source": [
    "## Allele level T-stat analysis of tap response data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbfecc89-cd0a-472b-bb6d-a2de6fc1a2b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "recovery_dura_allele=pd.DataFrame(columns = [\"dataset\",\"Recovery Duration\"])\n",
    "recovery_prob_allele=pd.DataFrame(columns = [\"dataset\",\"Recovery Probability\"])\n",
    "recovery_speed_allele=pd.DataFrame(columns = [\"dataset\",\"Recovery Speed\"])\n",
    "init_dura_allele=pd.DataFrame(columns = [\"dataset\",\"Initial Duration\"])\n",
    "init_prob_allele=pd.DataFrame(columns = [\"dataset\",\"Initial Probability\"])\n",
    "init_speed_allele=pd.DataFrame(columns = [\"dataset\",\"Initial Speed\"])\n",
    "final_dura_allele=pd.DataFrame(columns = [\"dataset\",\"Final Duration\"])\n",
    "final_prob_allele=pd.DataFrame(columns = [\"dataset\",\"Final Probability\"])\n",
    "final_speed_allele=pd.DataFrame(columns = [\"dataset\",\"Final Speed\"])\n",
    "hab_dura_allele=pd.DataFrame(columns = [\"dataset\",\"Habituation of Duration\"])\n",
    "hab_prob_allele=pd.DataFrame(columns = [\"dataset\",\"Habituation of Probability\"])\n",
    "hab_speed_allele=pd.DataFrame(columns = [\"dataset\",\"Habituation of Speed\"])\n",
    "\n",
    "list_tap_Tstats_allele = [recovery_dura_allele,\n",
    "                   recovery_prob_allele,\n",
    "                   recovery_speed_allele,\n",
    "                   init_dura_allele,\n",
    "                   init_prob_allele,\n",
    "                   init_speed_allele,\n",
    "                   final_dura_allele,\n",
    "                   final_prob_allele,\n",
    "                   final_speed_allele,\n",
    "                   hab_dura_allele,\n",
    "                   hab_prob_allele,\n",
    "                   hab_speed_allele]\n",
    "list_tap_metrics = [\"recovery_dura\",\n",
    "                    \"recovery_prob\",\n",
    "                    \"recovery_speed\",\n",
    "                    \"init_dura\",\n",
    "                    \"init_prob\",\n",
    "                    \"init_speed\",\n",
    "                    \"final_dura\",\n",
    "                    \"final_prob\",\n",
    "                    \"final_speed\",\n",
    "                    \"habit_dura\",\n",
    "                    \"habit_prob\",\n",
    "                    \"habit_speed\"]\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "440814af-d3d3-4972-a564-0743b642bc87",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in tap_data['dataset'].unique():\n",
    "    if x == \"N2\":\n",
    "    # if x == \"N2_N2\" or x == \"N2_XJ1\":\n",
    "        pass\n",
    "    else:\n",
    "        PD_habituation_allele=tap_data[tap_data['dataset']==x]\n",
    "        gene_PD_habituation_data=tap_data[tap_data['Date'].isin(PD_habituation_allele['Date'].unique())]\n",
    "        gene_PD_habituation_data_final=gene_PD_habituation_data[gene_PD_habituation_data['dataset'].isin(['N2', x])]\n",
    "        # gene_PD_habituation_data_final=gene_PD_habituation_data[gene_PD_habituation_data['dataset'].isin(['N2_N2', \"N2_XJ1\", x])]\n",
    "        # print(gene_baseline_data_final['Gene'].unique())\n",
    "        for a,b in zip(list_tap_metrics, list_tap_Tstats_allele):\n",
    "            TTest_Allele(a, gene_PD_habituation_data_final, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b91a2eb5-8d4f-4e4d-bebc-5ab154fd28fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_habituation_Tstats_allele=pd.DataFrame()\n",
    "for a in list_tap_Tstats_allele:\n",
    "    b=a.groupby(['dataset'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if b.columns.values[1] == list_tap_Tstats_allele[0].columns.values[1]:\n",
    "        PD_habituation_Tstats_allele=b\n",
    "    else:\n",
    "        PD_habituation_Tstats_allele=PD_habituation_Tstats_allele.join(b.iloc[:,1])\n",
    "        \n",
    "print(PD_habituation_Tstats_allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80468efc-200e-4a11-943d-18b5762996e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_habituation_Tstats_allele=PD_habituation_Tstats_allele.set_index(\"dataset\")\n",
    "print(PD_habituation_Tstats_allele)\n",
    "PD_habituation_Tstats_allele_sorted=PD_habituation_Tstats_allele.sort_index()\n",
    "print(PD_habituation_Tstats_allele_sorted)\n",
    "print(PD_baseline_Tstats_allele_sort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82f44a6d-6632-4e0f-8c58-a7060daf756d",
   "metadata": {},
   "source": [
    "## Gene-level T-stat analysis of Tap response data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "757bb7d2-d1e7-4227-a056-0ca0e1d10c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "recovery_dura=pd.DataFrame(columns = [\"Gene\",\"Recovery Duration\"])\n",
    "recovery_prob=pd.DataFrame(columns = [\"Gene\",\"Recovery Probability\"])\n",
    "recovery_speed=pd.DataFrame(columns = [\"Gene\",\"Recovery Speed\"])\n",
    "init_dura=pd.DataFrame(columns = [\"Gene\",\"Initial Duration\"])\n",
    "init_prob=pd.DataFrame(columns = [\"Gene\",\"Initial Probability\"])\n",
    "init_speed=pd.DataFrame(columns = [\"Gene\",\"Initial Speed\"])\n",
    "final_dura=pd.DataFrame(columns = [\"Gene\",\"Final Duration\"])\n",
    "final_prob=pd.DataFrame(columns = [\"Gene\",\"Final Probability\"])\n",
    "final_speed=pd.DataFrame(columns = [\"Gene\",\"Final Speed\"])\n",
    "hab_dura=pd.DataFrame(columns = [\"Gene\",\"Habituation of Duration\"])\n",
    "hab_prob=pd.DataFrame(columns = [\"Gene\",\"Habituation of Probability\"])\n",
    "hab_speed=pd.DataFrame(columns = [\"Gene\",\"Habituation of Speed\"])\n",
    "\n",
    "list_tap_Tstats = [recovery_dura,\n",
    "                   recovery_prob,\n",
    "                   recovery_speed,\n",
    "                   init_dura,\n",
    "                   init_prob,\n",
    "                   init_speed,\n",
    "                   final_dura,\n",
    "                   final_prob,\n",
    "                   final_speed,\n",
    "                   hab_dura,\n",
    "                   hab_prob,\n",
    "                   hab_speed]\n",
    "list_tap_metrics = [\"recovery_dura\",\n",
    "                    \"recovery_prob\",\n",
    "                    \"recovery_speed\",\n",
    "                    \"init_dura\",\n",
    "                    \"init_prob\",\n",
    "                    \"init_speed\",\n",
    "                    \"final_dura\",\n",
    "                    \"final_prob\",\n",
    "                    \"final_speed\",\n",
    "                    \"habit_dura\",\n",
    "                    \"habit_prob\",\n",
    "                    \"habit_speed\"]\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d4aa5c-fd22-4be8-99d7-f7d65ff0bc75",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in tap_data['Gene'].unique():\n",
    "    if x == \"N2\":\n",
    "        pass\n",
    "    else:\n",
    "        PD_habituation_gene=tap_data[tap_data['Gene']==x]\n",
    "        gene_PD_habituation_data=tap_data[tap_data['Date'].isin(PD_habituation_gene['Date'].unique())]\n",
    "        gene_PD_habituation_data_final=gene_PD_habituation_data[gene_PD_habituation_data['Gene'].isin(['N2', x])]\n",
    "        # print(gene_baseline_data_final['Gene'].unique())\n",
    "        for a,b in zip(list_tap_metrics, list_tap_Tstats):\n",
    "            TTest_Gene(a, gene_PD_habituation_data_final, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec7483b-d593-44b4-b450-dcb6a4f7866e",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_habituation_Tstats=pd.DataFrame()\n",
    "for a in list_tap_Tstats:\n",
    "    b=a.groupby(['Gene'], as_index=False).mean()\n",
    "    # print(a.columns.values[1])\n",
    "    if b.columns.values[1] == list_tap_Tstats[0].columns.values[1]:\n",
    "        PD_habituation_Tstats=b\n",
    "    else:\n",
    "        PD_habituation_Tstats=PD_habituation_Tstats.join(b.iloc[:,1])\n",
    "        \n",
    "print(PD_habituation_Tstats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad7d7dfd-2348-4c9c-97d7-c0c696d54ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "PD_habituation_Tstats=PD_habituation_Tstats.set_index(\"Gene\")\n",
    "PD_habituation_Tstats_sorted=PD_habituation_Tstats.sort_index()\n",
    "print(PD_habituation_Tstats)\n",
    "print(PD_habituation_Tstats_sorted)\n",
    "print(PD_baseline_Tstats_sort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3071bf3-afd5-49e8-baa4-7e37dd120383",
   "metadata": {},
   "source": [
    "# Merging t-stat data into one dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c23c436e-1fd2-4bd1-8dd4-53517d4d0f4a",
   "metadata": {},
   "source": [
    "## Gene-level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81eacd88-edb5-4e98-9f49-2d4508cf40c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_Tstats = pd.merge(PD_baseline_Tstats, PD_habituation_Tstats, on =['Gene'], how ='left')\n",
    "combined_Tstats=combined_Tstats.sort_index()\n",
    "# PD_combined_Tstats=PD_combined_Tstats.drop(index=\"N2\")\n",
    "# combined_Tstats=combined_Tstats.dropna()\n",
    "print(combined_Tstats)\n",
    "combined_Tstats_normalize=(combined_Tstats-combined_Tstats.mean())/combined_Tstats.std()\n",
    "print(combined_Tstats_normalize)\n",
    "print(combined_Tstats_normalize[combined_Tstats_normalize.index==\"N2\"])\n",
    "print(combined_Tstats_normalize.index)\n",
    "combined_Tstats_normalize_2=combined_Tstats_normalize-combined_Tstats_normalize[combined_Tstats_normalize.index==\"N2\"].squeeze()\n",
    "print(combined_Tstats_normalize_2[combined_Tstats_normalize_2.index==\"N2\"])\n",
    "\n",
    "first_col=combined_Tstats.pop(\"Area\")\n",
    "combined_Tstats.insert(0,\"Area\",first_col)\n",
    "\n",
    "first_col=combined_Tstats.pop(\"Length\")\n",
    "combined_Tstats.insert(0,\"Length\",first_col)\n",
    "\n",
    "first_col=combined_Tstats.pop(\"Width\")\n",
    "combined_Tstats.insert(0,\"Width\",first_col)\n",
    "\n",
    "first_col=combined_Tstats.pop(\"Angular Speed\")\n",
    "combined_Tstats.insert(5,\"Angular Speed\",first_col)\n",
    "\n",
    "print(combined_Tstats)\n",
    "\n",
    "combined_Tstats=combined_Tstats.rename(columns={\"Habituation of Duration\":\"Habituation of Response Duration\",\n",
    "                                                 \"Habituation of Probability\": \"Habituation of Respones Probability\",\n",
    "                                                 \"Habituation of Speed\":\"Habituation of Response Speed\",\n",
    "                                                 \"Initial Duration\": \"Initial Response Duration\",\n",
    "                                                 \"Initial Probability\": \"Initial Response Probability\",\n",
    "                                                 \"Initial Speed\": \"Initial Response Speed\",\n",
    "                                                 \"Final Duration\": \"Final Response Duration\",\n",
    "                                                 \"Final Probability\": \"Final Response Probability\",\n",
    "                                                 \"Final Speed\": \"Final Response Speed\",\n",
    "                                                 \"Recovery Duration\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                                 \"Recovery Probability\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                                 \"Recovery Speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "\n",
    "combined_Tstats_normalize_2=combined_Tstats_normalize_2.rename(columns={\"Habituation of Duration\":\"Habituation of Response Duration\",\n",
    "                                                                         \"Habituation of Probability\": \"Habituation of Respones Probability\",\n",
    "                                                                         \"Habituation of Speed\":\"Habituation of Response Speed\",\n",
    "                                                                         \"Initial Duration\": \"Initial Response Duration\",\n",
    "                                                                         \"Initial Probability\": \"Initial Response Probability\",\n",
    "                                                                         \"Initial Speed\": \"Initial Response Speed\",\n",
    "                                                                         \"Final Duration\": \"Final Response Duration\",\n",
    "                                                                         \"Final Probability\": \"Final Response Probability\",\n",
    "                                                                         \"Final Speed\": \"Final Response Speed\",\n",
    "                                                                         \"Recovery Duration\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                                                         \"Recovery Probability\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                                                         \"Recovery Speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "print(combined_Tstats_normalize_2)\n",
    "\n",
    "first_col=combined_Tstats_normalize_2.pop(\"Area\")\n",
    "combined_Tstats_normalize_2.insert(0,\"Area\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_2.pop(\"Length\")\n",
    "combined_Tstats_normalize_2.insert(0,\"Length\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_2.pop(\"Width\")\n",
    "combined_Tstats_normalize_2.insert(0,\"Width\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_2.pop(\"Angular Speed\")\n",
    "combined_Tstats_normalize_2.insert(5,\"Angular Speed\",first_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_2.pop(\"Spontaneous Recovery of Response Duration\")\n",
    "combined_Tstats_normalize_2.insert(22,\"Spontaneous Recovery of Response Duration\",last_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_2.pop(\"Spontaneous Recovery of Response Probability\")\n",
    "combined_Tstats_normalize_2.insert(22,\"Spontaneous Recovery of Response Probability\",last_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_2.pop(\"Spontaneous Recovery of Response Speed\")\n",
    "combined_Tstats_normalize_2.insert(22,\"Spontaneous Recovery of Response Speed\",last_col)\n",
    "\n",
    "print(combined_Tstats_normalize_2)\n",
    "\n",
    "combined_Tstats_melted=combined_Tstats.reset_index()\n",
    "combined_Tstats_melted=pd.melt(combined_Tstats_melted, id_vars=['Gene'],\n",
    "                               var_name='Metric',\n",
    "                               value_name='T_score')\n",
    "combined_Tstats_melted_sorted=combined_Tstats_melted.sort_values(by=['T_score'])\n",
    "print(combined_Tstats_melted_sorted)\n",
    "\n",
    "combined_Tstats_normalized_melted=combined_Tstats_normalize_2.reset_index()\n",
    "combined_Tstats_normalized_melted=pd.melt(combined_Tstats_normalized_melted, id_vars=['Gene'],\n",
    "                                           var_name='Metric',\n",
    "                                           value_name='T_score')\n",
    "print(combined_Tstats_normalized_melted)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73da0e1e-d702-408f-b278-2a05b39ba999",
   "metadata": {},
   "source": [
    "## Allele level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16521a12-8c86-4310-9455-09054ed08b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_Tstats_allele = pd.merge(PD_baseline_Tstats_allele, PD_habituation_Tstats_allele, on =['dataset'], how ='left')\n",
    "combined_Tstats_allele=combined_Tstats_allele.sort_index()\n",
    "# PD_combined_Tstats=PD_combined_Tstats.drop(index=\"N2\")\n",
    "# combined_Tstats=combined_Tstats_allele.dropna()\n",
    "print(\"combined_Tstats_allele\")\n",
    "print(combined_Tstats_allele)\n",
    "print(\"-\"*55)\n",
    "combined_Tstats_normalize_allele=(combined_Tstats_allele-combined_Tstats_allele.mean())/combined_Tstats_allele.std()\n",
    "print(\"combined_Tstats_normalize_allele\")\n",
    "print(combined_Tstats_normalize_allele)\n",
    "print(\"-\"*55)\n",
    "# print(combined_Tstats_normalize_allele[combined_Tstats_normalize_allele.index==\"N2\"])\n",
    "print(\"N2_XJ1 Data\")\n",
    "print(combined_Tstats_normalize_allele[combined_Tstats_normalize_allele.index==\"N2_XJ1\"])\n",
    "print(\"-\"*55)\n",
    "print(\"combined_Tstats_normalize_allele row labels\")\n",
    "print(combined_Tstats_normalize_allele.index)\n",
    "print(\"-\"*55)\n",
    "# combined_Tstats_normalize_allele_2=combined_Tstats_normalize_allele-combined_Tstats_normalize_allele[combined_Tstats_normalize_allele.index==\"N2\"].squeeze()\n",
    "combined_Tstats_normalize_allele_2=combined_Tstats_normalize_allele-combined_Tstats_normalize_allele[combined_Tstats_normalize_allele.index==\"N2_XJ1\"].squeeze()\n",
    "# print(combined_Tstats_normalize_allele_2[combined_Tstats_normalize_allele_2.index==\"N2\"])\n",
    "print(\"combined Tstats normalized N2_XJ1 data\")\n",
    "print(combined_Tstats_normalize_allele_2[combined_Tstats_normalize_allele_2.index==\"N2_XJ1\"])\n",
    "print(\"-\"*55)\n",
    "first_col=combined_Tstats_allele.pop(\"Area\")\n",
    "combined_Tstats_allele.insert(0,\"Area\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_allele.pop(\"Length\")\n",
    "combined_Tstats_allele.insert(0,\"Length\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_allele.pop(\"Width\")\n",
    "combined_Tstats_allele.insert(0,\"Width\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_allele.pop(\"Angular Speed\")\n",
    "combined_Tstats_allele.insert(5,\"Angular Speed\",first_col)\n",
    "print(\"combined_Tstats_allele\")\n",
    "print(combined_Tstats_allele)\n",
    "print(\"-\"*55)\n",
    "combined_Tstats_allele=combined_Tstats_allele.rename(columns={\"Habituation of Duration\":\"Habituation of Response Duration\",\n",
    "                                                             \"Habituation of Probability\": \"Habituation of Respones Probability\",\n",
    "                                                             \"Habituation of Speed\":\"Habituation of Response Speed\",\n",
    "                                                             \"Initial Duration\": \"Initial Response Duration\",\n",
    "                                                             \"Initial Probability\": \"Initial Response Probability\",\n",
    "                                                             \"Initial Speed\": \"Initial Response Speed\",\n",
    "                                                             \"Final Duration\": \"Final Response Duration\",\n",
    "                                                             \"Final Probability\": \"Final Response Probability\",\n",
    "                                                             \"Final Speed\": \"Final Response Speed\",\n",
    "                                                             \"Recovery Duration\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                                             \"Recovery Probability\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                                             \"Recovery Speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "\n",
    "combined_Tstats_normalize_allele_2=combined_Tstats_normalize_allele_2.rename(columns={\"Habituation of Duration\":\"Habituation of Response Duration\",\n",
    "                                                                                             \"Habituation of Probability\": \"Habituation of Respones Probability\",\n",
    "                                                                                             \"Habituation of Speed\":\"Habituation of Response Speed\",\n",
    "                                                                                             \"Initial Duration\": \"Initial Response Duration\",\n",
    "                                                                                             \"Initial Probability\": \"Initial Response Probability\",\n",
    "                                                                                             \"Initial Speed\": \"Initial Response Speed\",\n",
    "                                                                                             \"Final Duration\": \"Final Response Duration\",\n",
    "                                                                                             \"Final Probability\": \"Final Response Probability\",\n",
    "                                                                                             \"Final Speed\": \"Final Response Speed\",\n",
    "                                                                                             \"Recovery Duration\": \"Spontaneous Recovery of Response Duration\",\n",
    "                                                                                             \"Recovery Probability\": \"Spontaneous Recovery of Response Probability\",\n",
    "                                                                                             \"Recovery Speed\": \"Spontaneous Recovery of Response Speed\"})\n",
    "\n",
    "print(\"combined_Tstats_normalize_allele_2\")\n",
    "print(combined_Tstats_normalize_allele_2)\n",
    "print(\"-\"*55)\n",
    "first_col=combined_Tstats_normalize_allele_2.pop(\"Area\")\n",
    "combined_Tstats_normalize_allele_2.insert(0,\"Area\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_allele_2.pop(\"Length\")\n",
    "combined_Tstats_normalize_allele_2.insert(0,\"Length\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_allele_2.pop(\"Width\")\n",
    "combined_Tstats_normalize_allele_2.insert(0,\"Width\",first_col)\n",
    "\n",
    "first_col=combined_Tstats_normalize_allele_2.pop(\"Angular Speed\")\n",
    "combined_Tstats_normalize_allele_2.insert(5,\"Angular Speed\",first_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_allele_2.pop(\"Spontaneous Recovery of Response Duration\")\n",
    "combined_Tstats_normalize_allele_2.insert(22,\"Spontaneous Recovery of Response Duration\",last_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_allele_2.pop(\"Spontaneous Recovery of Response Probability\")\n",
    "combined_Tstats_normalize_allele_2.insert(22,\"Spontaneous Recovery of Response Probability\",last_col)\n",
    "\n",
    "last_col=combined_Tstats_normalize_allele_2.pop(\"Spontaneous Recovery of Response Speed\")\n",
    "combined_Tstats_normalize_allele_2.insert(22,\"Spontaneous Recovery of Response Speed\",last_col)\n",
    "\n",
    "print(combined_Tstats_normalize_allele_2)\n",
    "\n",
    "combined_Tstats_normalized_melted_allele=combined_Tstats_normalize_allele_2.reset_index()\n",
    "combined_Tstats_normalized_melted_allele=pd.melt(combined_Tstats_normalized_melted_allele, id_vars=['dataset'],\n",
    "                                                   var_name='Metric',\n",
    "                                                   value_name='T_score')\n",
    "print(combined_Tstats_normalized_melted_allele)\n",
    "\n",
    "\n",
    "combined_Tstats_melted_allele=combined_Tstats_allele.reset_index()\n",
    "combined_Tstats_melted_allele=pd.melt(combined_Tstats_melted_allele, id_vars=['dataset'],\n",
    "                                       var_name='Metric',\n",
    "                                       value_name='T_score')\n",
    "print(combined_Tstats_melted_allele)\n",
    "combined_Tstats_melted_allele_sorted=combined_Tstats_melted_allele.sort_values(by=['T_score'])\n",
    "print(combined_Tstats_melted_allele_sorted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb957ef0-6661-492a-8d1e-d3d5cff2dc3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_Tstats_normalize_2['Screen']=Screen\n",
    "combined_Tstats_normalize_allele_2['Screen']=Screen\n",
    "combined_Tstats_normalized_melted['Screen']=Screen\n",
    "combined_Tstats_normalized_melted_allele['Screen']=Screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe9b29dd-6dbf-411f-b05d-b4456fe16bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tap_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4be480f8-c52d-4dac-a6e5-2c5bd003001c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(combined_MSD)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4915fd6-beba-45a7-9575-346dfe679285",
   "metadata": {},
   "source": [
    "## sqlite3 upload data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d4a06cb-457b-4d9e-bb11-6e6042b6b65b",
   "metadata": {},
   "source": [
    "## A janky way to add data and update the sql \n",
    "\n",
    "1. Read table to pd.DataFrame\n",
    "2. Add new data to pd.DataFrame\n",
    "3. Replace old table with newly updated pd.DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375b2702-0d88-4fc8-8b4d-fc5c633dee1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read(table):\n",
    "    result = pd.read_sql_query(f\"SELECT * FROM {table}\", conn)\n",
    "    return result\n",
    "\n",
    "conn = sqlite3.connect('/Users/Joseph/Desktop/NRSC510B/mwt_data.db')\n",
    "old_tap_output = read('tap_response_data')\n",
    "old_tap_tstat_gene = read('tstat_gene_data')\n",
    "old_tap_tstat_allele = read('tstat_allele_data')\n",
    "old_gene_profile_data = read('gene_profile_data')\n",
    "old_allele_profile_data = read('allele_profile_data')\n",
    "# old_gene_MSD = read('gene_MSD')\n",
    "# old_allele_MSD = read('allele_MSD')\n",
    "\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d628dd4-08e4-4e28-bfb7-33814ed81633",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(old_tap_tstat_allele)\n",
    "# # print(old_tap_tstat_gene.drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31b3c1d0-318c-40f6-b6a3-e5e23b781539",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # conn=sqlite3.connect('/Users/Joseph/Desktop/NRSC510B/mwt_data.db')\n",
    "# old_tap_output.drop_duplicates().to_sql('tap_response_data', conn, if_exists='replace', index=False)\n",
    "# # baseline_output.to_sql('tap_baseline_data', conn, if_exists='replace', index=False)\n",
    "# old_tap_tstat_gene.drop_duplicates().to_sql('tstat_gene_data', conn, if_exists='replace', index=False)\n",
    "# old_tap_tstat_allele.drop_duplicates().to_sql('tstat_allele_data', conn, if_exists='replace', index=False)\n",
    "# old_gene_profile_data.drop_duplicates().to_sql('gene_profile_data', conn, if_exists='replace', index=False)\n",
    "# old_allele_profile_data.drop_duplicates().to_sql('allele_profile_data', conn, if_exists='replace', index=False)\n",
    "# # combined_MSD.to_sql('gene_MSD', conn, if_exists='append', index=False)\n",
    "# # allele_combined_MSD.to_sql('allele_MSD', conn, if_exists='append', index=False)\n",
    "# # combined_Tstats_melted_sorted.to_sql('allele_phenotype_data', conn, if_exists='replace', index=False)\n",
    "# print(conn.total_changes)\n",
    "# conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3030034-a9bd-44d9-95ff-32b11581adf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "conn = sqlite3.connect('/Users/Joseph/Desktop/NRSC510B/mwt_data.db')\n",
    "cur = conn.cursor()\n",
    "cur.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "print(cur.fetchall())\n",
    "# cur.execute(\"DROP TABLE IF EXISTS tap_baseline_data\")\n",
    "# cur.execute(\"DROP TABLE IF EXISTS allele_phenotype_data\")\n",
    "# cur.execute(\"DROP TABLE IF EXISTS tstat_data\")\n",
    "\n",
    "# print(\"-\"*60)\n",
    "\n",
    "# cur.execute(\"SELECT name FROM sqlite_master WHERE type='table';\")\n",
    "print(cur.fetchall())\n",
    "# cur.execute(\".tables\")\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0e7cd78-fd93-4a05-b56c-406ea0c688af",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "input: (pd.DataFrame, pd.Dataframe or None, str)\n",
    "    1 or 2 pd.DataFrames with the same indexes (columns), and sql table name (str)\n",
    "    first DataFrame is old DataFrame read from the .db file\n",
    "    second DataFrame is new processed DataFrame to be appended to the old one\n",
    "    \n",
    "\n",
    "function: If only one DataFrame, replace the .db tabl with this one DataFrame\n",
    " Otherwise (when given two dataframes):\n",
    "     1. concatenate new DataFrame to old DataFrame\n",
    "     2. to_sql, if_exists = 'replace'\n",
    "\n",
    "'''\n",
    "\n",
    "def update_table(df1, df2=None, table_name):\n",
    "    if df2=None:\n",
    "        df1.to_sql(table_name, conn, if_exists='replace', index=False)\n",
    "    else:\n",
    "        new_df=pd.concat([df1, df2], ignore_index=True)\n",
    "        new_df.to_sql(table_name, conn, if_exists='replace', index=False)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d4512d-bb07-466e-892d-fb0c99bc25b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tap_output\n",
    "combined_Tstats_normalize_2\n",
    "combined_Tstats_normalize_allele_2\n",
    "combined_Tstats_normalized_melted\n",
    "combined_Tstats_normalized_melted_allele\n",
    "combined_MSD\n",
    "allele_combined_MSD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d33c616-2839-4572-af70-233af783302a",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn=sqlite3.connect('/Users/Joseph/Desktop/NRSC510B/mwt_data.db')\n",
    "tap_output.to_sql('tap_response_data', conn, if_exists='append', index=False)\n",
    "# baseline_output.to_sql('tap_baseline_data', conn, if_exists='replace', index=False)\n",
    "combined_Tstats_normalize_2.reset_index().to_sql('tstat_gene_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalize_allele_2.reset_index().to_sql('tstat_allele_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalized_melted.to_sql('gene_profile_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalized_melted_allele.to_sql('allele_profile_data', conn, if_exists='append', index=False)\n",
    "combined_MSD.to_sql('gene_MSD', conn, if_exists='append', index=False)\n",
    "allele_combined_MSD.to_sql('allele_MSD', conn, if_exists='append', index=False)\n",
    "# combined_Tstats_melted_sorted.to_sql('allele_phenotype_data', conn, if_exists='replace', index=False)\n",
    "print(conn.total_changes)\n",
    "conn.close()\n",
    "\n",
    "# Want to test edge cases of pd.to_sql functionality#############"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90d4822f-dd6f-43eb-98b5-41b7190223f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn=sqlite3.connect('/Users/Joseph/Desktop/NRSC510B/mwt_data.db')\n",
    "tap_output.to_sql('tap_response_data', conn, if_exists='append', index=False)\n",
    "# baseline_output.to_sql('tap_baseline_data', conn, if_exists='replace', index=False)\n",
    "combined_Tstats_normalize_2.reset_index().to_sql('tstat_gene_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalize_allele_2.reset_index().to_sql('tstat_allele_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalized_melted.to_sql('gene_profile_data', conn, if_exists='append', index=False)\n",
    "combined_Tstats_normalized_melted_allele.to_sql('allele_profile_data', conn, if_exists='append', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0914e50b-09ea-4714-9bb5-5b0bb25924b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(conn.total_changes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d2c5b51-c902-4ae5-8506-1fbd49101e69",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursor = conn.cursor()\n",
    "\n",
    "cursor.execute(\"SELECT COUNT(*) FROM table_name WHERE primary_key_column = ?\", (value,))\n",
    "exists = cursor.fetchone()[0] > 0\n",
    "\n",
    "if exists:\n",
    "    print(\"Primary key already exists in the table.\")\n",
    "else:\n",
    "    print(\"Primary key does not exist in the table.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc7e5100-bdca-44d5-bd0c-84b28e11cbcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "700ed37c-240d-41d2-b90a-8793a23477b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7611f10-536b-4cd4-a402-01aa76c5af69",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e10398c8-2680-4ce2-a39a-d348d8246943",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b65df115-ff9a-4ad1-811f-df33002845cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66abb924-f1af-4cdd-b99a-f59d9255da5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7895c9a9-f067-495c-a9ff-6a50a4887ff1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6e16d206-c8af-4ddf-8412-a86118024cec",
   "metadata": {},
   "source": [
    "## Clustering Playtest\n",
    "\n",
    "The below cells are for exploring clustering functionalities. Currently under development."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2079074-5dcb-45aa-84c6-e1504f75a2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc0ffccd-4417-4253-9403-57bdbb77902b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca.transform(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94edc1ff-abed-494e-af6e-f0968a6f5555",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca=PCA()\n",
    "pca.fit(X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbf7fb34-82cd-43e9-8528-2e5f2bca26f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "components=pca.transform(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c07dbc-21ba-44ad-a272-d790d6629eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(components)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b771ba47-bc21-41f5-9fbd-3f0127b37251",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "\n",
    "features = [\"Instantaneous Speed\", \"Interval Speed\", \"Bias\", \"Width\",\"Length\"]\n",
    "\n",
    "pca = PCA()\n",
    "components = pca.fit_transform(X[features])\n",
    "\n",
    "labels = {\n",
    "    str(i): f\"PC {i+1} ({var:.1f}%)\"\n",
    "    for i, var in enumerate(pca.explained_variance_ratio_ * 100)\n",
    "}\n",
    "\n",
    "fig = px.scatter_matrix(\n",
    "    components,width=800, height=800,\n",
    "    labels=labels,\n",
    "    dimensions=range(4))\n",
    "    # color=df[\"species\"]\n",
    "# fig.update_traces(diagonal_visible=False)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b76c3485-f5fd-475e-ad6c-757e0eda810e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7289b1cf-f489-472c-a4e6-60a7cc5d5dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(old_tap_tstat_allele)\n",
    "\n",
    "\n",
    "meta=old_tap_tstat_allele.dropna()[['dataset','Screen']]\n",
    "X=old_tap_tstat_allele.dropna().drop(columns=['dataset','Screen'])\n",
    "# print(meta)\n",
    "# print(X)\n",
    "\n",
    "print(\"-\"*55)\n",
    "filtered = old_tap_tstat_allele[old_tap_tstat_allele['Screen']==\"PD_Screen\"]\n",
    "filtered_meta = filtered.dropna()[['dataset','Screen']]\n",
    "filtered_X = filtered.dropna().drop(columns=['dataset','Screen'])\n",
    "\n",
    "# print(filtered_X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb4ffdc-22c9-427e-a9c1-5c0c762ee404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(old_tap_tstat_allele)\n",
    "import numpy as np\n",
    "filtered=old_tap_tstat_allele.drop(columns=['Spontaneous Recovery of Response Duration',\n",
    "                                          'Spontaneous Recovery of Response Probability','Spontaneous Recovery of Response Speed'])\n",
    "filtered_no_na=filtered.dropna()\n",
    "print(filtered_no_na)\n",
    "meta=filtered_no_na[['dataset','Screen']]\n",
    "X = filtered_no_na.drop(columns=['dataset','Screen'])\n",
    "print(np.unique(filtered_no_na.Screen))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65da8603-2cbc-4ba0-bd49-310a124193f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "figure=sns.pairplot(old_tap_tstat_gene.drop(\"Gene\", axis=1), hue='Screen')\n",
    "figure.savefig('/Users/Joseph/Desktop/MWT_Datasets/pairpot.png', bbox_inches=\"tight\", dpi=450) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79bb657e-8fb0-4085-b59f-f4676282be5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import umap\n",
    "reducer = umap.UMAP()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80149b5c-98b9-444a-8d7e-ac5cebbb4398",
   "metadata": {},
   "outputs": [],
   "source": [
    "embedding = reducer.fit_transform(X)\n",
    "embedding.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ee9c89-6ff5-4eec-b55b-3d0dd9cd1849",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = umap.UMAP(\n",
    "        n_neighbors=10,\n",
    "        min_dist=0.1,\n",
    "        metric='euclidean'\n",
    "        # n_components=n_components,\n",
    "        # metric=metric\n",
    ")\n",
    "u = fit.fit_transform(X)\n",
    "u.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4621dd0a-42af-4fb6-8411-163782039998",
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = umap.UMAP(\n",
    "        n_neighbors=5,\n",
    "        min_dist=0.1,\n",
    "        metric='euclidean'\n",
    "        # n_components=n_components,\n",
    "        # metric=metric\n",
    ")\n",
    "filtered_u = fit.fit_transform(filtered_X)\n",
    "filtered_u.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f229886-4063-4f49-ad85-3b45bcf8cfa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "plt.scatter(\n",
    "    u[:, 0],\n",
    "    u[:, 1],\n",
    "    c=[sns.color_palette()[x] for x in filtered_no_na.dropna().Screen.map({\"PD_Screen\":0, \"ASD_Screen\":1, \"Glia_Genes_Screen\":2, \"G-Proteins_Screen\":3, \"Neuron_Genes_Screen\":4})],\n",
    "    )\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.gca().set_aspect('equal', 'datalim')\n",
    "plt.title('UMAP projection of the MWT dataset', fontsize=24);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddeb931-0f03-4f56-aec7-dd85748adf92",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.scatter(\n",
    "    filtered_u[:, 0],\n",
    "    filtered_u[:, 1],\n",
    "    c=[sns.color_palette()[x] for x in filtered.dropna().Screen.map({\"PD_Screen\":0, \"ASD_Screen\":1, \"Glia_Genes_Screen\":2})])\n",
    "plt.gca().set_aspect('equal', 'datalim')\n",
    "plt.title('UMAP projection of the MWT dataset', fontsize=24);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "036c0fd2-c463-4163-9cf1-b4d3ab65d547",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
